<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="google-translate-customization" content="108d9124921d80c3-80e20d618ff053c8-g4f02ec6f3dba68b7-c"></meta><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Scanorama | UCI genPALS</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Scanorama" />
<meta name="author" content="Axel Almet" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Author: Axel Almet" />
<meta property="og:description" content="Author: Axel Almet" />
<link rel="canonical" href="/integration/2020/12/03/scanorama_demo_pancreas.html" />
<meta property="og:url" content="/integration/2020/12/03/scanorama_demo_pancreas.html" />
<meta property="og:site_name" content="UCI genPALS" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2020-12-03T00:00:00-08:00" />
<script type="application/ld+json">
{"headline":"Scanorama","url":"/integration/2020/12/03/scanorama_demo_pancreas.html","dateModified":"2020-12-03T00:00:00-08:00","datePublished":"2020-12-03T00:00:00-08:00","mainEntityOfPage":{"@type":"WebPage","@id":"/integration/2020/12/03/scanorama_demo_pancreas.html"},"author":{"@type":"Person","name":"Axel Almet"},"description":"Author: Axel Almet","@type":"BlogPosting","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="shortcut icon" href="">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-noto-sans@0.0.72/index.min.css">
  <link rel="stylesheet"
        href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/default.min.css">
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js"></script>
  <!-- and it's easy to individually load additional languages -->
  <script charset="UTF-8"
          src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/languages/go.min.js"></script>
  <link rel="stylesheet" href="/assets/css/main.css">
  <script src="/assets/js/main.js"></script><link type="application/atom+xml" rel="alternate" href="/feed.xml" title="UCI genPALS" /></head>
<body>



















<header class="site-header site-header-transparent" role="banner">

  <div class="wrapper">
    <div class="site-header-inner"><span class="site-brand"><a class="site-brand-inner" rel="author" href="/">
  <img class="site-favicon" title="UCI genPALS" src="" onerror="this.style.display='none'">
  UCI genPALS
</a>
</span><nav class="site-nav">
          <input type="checkbox" id="nav-trigger" class="nav-trigger" />
          <label for="nav-trigger">
            <span class="menu-icon">
              <svg viewBox="0 0 18 15" width="18px" height="15px">
                <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
              </svg>
            </span>
          </label>

          <div class="trigger"><a class="page-link" href="/about.html">ABOUT</a><a class="page-link" href="/archives.html">ARCHIVES</a><a class="page-link" href="/categories.html">CATEGORIES</a><a class="page-link" href="/events.html">EVENTS</a><a class="page-link" href="/">HOME</a><a class="page-link" href="/resources.html">RESOURCES</a><a class="page-link" href="/symposium.html">GENPALS SYMPOSIUM 2021</a><a class="page-link" href="/tags.html">TAGS</a>




<span class="page-link">

<div id="google_translate_element" style="display: none;">
</div>

<span class="ct-language">
  <ul class="list-unstyled ct-language-dropdown">
    
      <li>
        <a href="#" class="lang-select" data-lang="en">
          
          <img src="https://www.countryflags.io/us/flat/64.png" title="English">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="fr">
          
          <img src="https://www.countryflags.io/fr/flat/64.png" title="Franch">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="zh-CN">
          
          <img src="https://www.countryflags.io/cn/flat/64.png" title="Chinese(Simple)">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="ja">
          
          <img src="https://www.countryflags.io/jp/flat/64.png" title="Japan">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="ko">
          
          <img src="https://www.countryflags.io/kr/flat/64.png" title="Korean">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="ru">
          
          <img src="https://www.countryflags.io/ru/flat/64.png" title="Russia">
          
        </a>
      </li>
    
  </ul>
</span>

<script type="text/javascript">
function googleTranslateElementInit() {
  new google.translate.TranslateElement({
    pageLanguage: '',
    autoDisplay: false,
    layout: google.translate.TranslateElement.InlineLayout.VERTICAL
  }, 'google_translate_element');

  function restoreLang() {
    var iframe = document.getElementsByClassName('goog-te-banner-frame')[0];
    if (!iframe) return;

    var innerDoc = iframe.contentDocument || iframe.contentWindow.document;
    var restore_el = innerDoc.getElementsByTagName("button");

    for (var i = 0; i < restore_el.length; i++) {
      if (restore_el[i].id.indexOf("restore") >= 0) {
        restore_el[i].click();
        var close_el = innerDoc.getElementsByClassName("goog-close-link");
        close_el[0].click();
        return;
      }
    }
  }

  function triggerHtmlEvent(element, eventName) {
    var event;
    if (document.createEvent) {
      event = document.createEvent('HTMLEvents');
      event.initEvent(eventName, true, true);
      element.dispatchEvent(event);
    } else {
      event = document.createEventObject();
      event.eventType = eventName;
      element.fireEvent('on' + event.eventType, event);
    }
  }

  var googleCombo = document.querySelector("select.goog-te-combo");
  var langSelect = document.querySelector('.ct-language');
  langSelect.addEventListener('click', function(event) {
    if (!event.target) {
      return;
    }

    var selected = document.querySelector('.ct-language .ct-language-selected');
    if (selected) {
      selected.classList.remove('ct-language-selected');
    }

    var target = event.target;
    while (target && target !== langSelect ) {
      if (target.matches('.lang-select')) {
        break;
      }
      target = target.parentElement;
    }

    if (target && target.matches('.lang-select')) {
      var lang = target.getAttribute('data-lang');
      if (googleCombo.value == lang) {
        restoreLang();
      } else {
        target.parentElement.classList.add('ct-language-selected');
        googleCombo.value = lang;
        triggerHtmlEvent(googleCombo, 'change');
      }
    }

    event.preventDefault();
  });
}
</script>

<script type="text/javascript" src="//translate.google.com/translate_a/element.js?cb=googleTranslateElementInit"></script>
</span></div>
        </nav></div>
  </div>
</header>

<script>
  (function() {
    var lastScrollY = getScrollPos().y;
    var documentElement = document.documentElement;

    function storeScrollData() {
      var y = getScrollPos().y;documentElement.setAttribute("data-header-transparent", "");var scrollStatus = "";

      if (y <= 0) {
        scrollStatus = "top";
      } else if ((window.innerHeight + y) >= document.body.offsetHeight) {
        scrollStatus = "bottom";
      } else {
        var isScrollDown = (y - lastScrollY > 0) ? true : false;
        scrollStatus = isScrollDown ? "down" : "up";
      }

      lastScrollY = y;
      documentElement.setAttribute("data-scroll-status", scrollStatus);
    }

    window.addEventListener('scroll', function(e) {
      storeScrollData();
    });

    storeScrollData();
  })();
</script>











































  <section class="page-banner">
    <div class="page-banner-img">
      <div style="background-image: url(/assets/images/scanorama/scanorama_abstract.jpg)"></div>
    </div>
    <div class="wrapper">
      <div class="page-banner-inner"><header class="post-header">
  <h1 class="post-title p-name" itemprop="name headline">Scanorama</h1>
  <h3 class="post-subtitle">using Scanorama to integrate and batch correct scRNA-seq data</h3>

  <p class="post-meta">
    <time class="dt-published" datetime="2020-12-03T00:00:00-08:00" itemprop="datePublished"><i class="fa fa-calendar"></i> Dec 03, 2020
    </time>

    





















    <span class="post-reading-time left-vsplit"><i class="fa fa-clock-o"></i> About 11 mins</span>
  </p><div class="post-tags"><a class="post-tag" href="/tags.html#single-cell">#single-cell</a><a class="post-tag" href="/tags.html#batch-correction">#batch-correction</a><a class="post-tag" href="/tags.html#integration">#integration</a><a class="post-tag" href="/tags.html#python">#python</a><a class="post-tag" href="/tags.html#scanpy">#scanpy</a></div></header>
</div>
    </div>
  </section><script>
  function hashLocate(hashValue) {
    hashValue = hashValue.replace(/^.*#h-/, '');
    var element = document.getElementById(hashValue);

    if (!element) {
      return;
    }

    var header = document.querySelector('header.site-header');
    var headerRect = header.getBoundingClientRect();
    var headerTop = Math.floor(headerRect.top);
    var headerHeight = Math.floor(headerRect.height);
    var scrollPos = getScrollPos();
    var offsetY = element.offsetTop - (headerTop + headerHeight + 20);

    if (offsetY == scrollPos.y) {
      return;
    }

    if (headerTop == 0  && offsetY > scrollPos.y) {
      offsetY += headerHeight + 2;
    } else if (headerTop < 0  && offsetY < scrollPos.y) {
      offsetY -= headerHeight - 2;
    }

    smoothScrollTo(offsetY);
  }

  // The first event occurred
  window.addEventListener('load', function(event) {
    if (window.location.hash) {
      hashLocate(window.location.hash);
    }
  });

  // The first event occurred
  window.addEventListener('click', function(event) {
    if (event.target.tagName.toLowerCase() == 'a') {
      hashLocate(event.target.getAttribute('href'));
    }
  });
</script>
<div class="theme-toggle">
  <input type="checkbox" id="theme-switch">
  <label for="theme-switch">
    <div class="toggle"></div>
    <div class="names">
      <p class="light">Light</p>
      <p class="dark">Dark</p>
    </div>
  </label>
</div>




<script>
  (function() {
    var sw = document.getElementById('theme-switch');
    var html = document.getElementsByTagName('html')[0];
    var nightModeOption = ('on' || 'auto').toLowerCase();
    var storage = nightModeOption === 'manual'
        ? localStorage
        : sessionStorage;
    var themeData = loadThemeData();

    function saveThemeData(data) {
      storage.setItem('theme', JSON.stringify(data));
    }

    function loadThemeData() {
      var data = storage.getItem('theme');
      try {
        data = JSON.parse(data ? data : '');
      } catch(e) {
        data = { nightShift: undefined, autoToggleAt: 0 };
        saveThemeData(data);
      }
      return data;
    }

    function handleThemeToggle(nightShift) {
      themeData.nightShift = nightShift;
      saveThemeData(themeData);
      html.dataset.theme = nightShift ? 'dark' : 'light';
      setTimeout(function() {
        sw.checked = nightShift ? true : false;
      }, 50);
    }

    function autoThemeToggle() {
      // Next time point of theme toggle
      var now = new Date();
      var toggleAt = new Date();
      var hours = now.getHours();
      var nightShift = hours >= 19 || hours <=7;

      if (nightShift) {
        if (hours > 7) {
          toggleAt.setDate(toggleAt.getDate() + 1);
        }
        toggleAt.setHours(7);
      } else {
        toggleAt.setHours(19);
      }

      toggleAt.setMinutes(0);
      toggleAt.setSeconds(0);
      toggleAt.setMilliseconds(0)

      var delay = toggleAt.getTime() - now.getTime();

      // auto toggle theme mode
      setTimeout(function() {
        handleThemeToggle(!nightShift);
      }, delay);

      return {
        nightShift: nightShift,
        toggleAt: toggleAt.getTime()
      };
    }

    // Listen the theme toggle event
    sw.addEventListener('change', function(event) {
      handleThemeToggle(event.target.checked);
    });

    if (nightModeOption == 'auto') {
      var data = autoThemeToggle();

      // Toggle theme by local setting
      if (data.toggleAt > themeData.autoToggleAt) {
        themeData.autoToggleAt = data.toggleAt;
        handleThemeToggle(data.nightShift);
      } else {
        handleThemeToggle(themeData.nightShift);
      }
    } else if (nightModeOption == 'manual') {
      handleThemeToggle(themeData.nightShift);
    } else {
      var nightShift = themeData.nightShift;
      if (nightShift === undefined) {
        nightShift = nightModeOption === 'on';
      }
      handleThemeToggle(nightShift);
    }
  })();
</script>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <div class="framework">
  <section class="main">

     <div class="post">
  <section>









<article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

    <div class="post-content e-content" itemprop="articleBody">

      <p>Author: Axel Almet</p>

<h2 id="introduction-installation-and-data-download">introduction, installation, and data download</h2>

<p>Here, we will integrate four datasets of human pancreas cells from different studies. This collection has been used to benchmark a number of integration methods now and is used in the following <a href="https://scanpy-tutorials.readthedocs.io/en/latest/integrating-data-using-ingest.html">Scanpy tutorial</a>.</p>

<p>It’s also worth mentioning that a lot of what I will run here is per the guidelines of the benchmarking comparison by <a href="https://www.biorxiv.org/content/10.1101/2020.05.22.111161v1.abstract">Luecken et al. (2020)</a>. The code that was used to obtain the results in Luecken et al. can be found at the following <a href="https://github.com/theislab/scib">GitHub repository</a>.</p>

<p>First thing’s first, let’s install Scanpy, Scanorama, and, if you’d like to compare Scanorama’s performance with a similar method, BBKNN.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="n">scanpy</span>
<span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="n">scanorama</span>
<span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="n">bbknn</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># The following packages should already be part of Google Colba
</span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="n">sc</span>
<span class="kn">import</span> <span class="nn">scanorama</span> <span class="k">as</span> <span class="n">scrama</span>
</code></pre></div></div>

<p>Obtain the Pancreas dataset used in the Scanpy tutorial.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># If you want to see how many cells there are:
</span><span class="n">pancreas_all</span> <span class="o">=</span> <span class="n">sc</span><span class="p">.</span><span class="n">read</span><span class="p">(</span><span class="s">'data/pancreas.h5ad'</span><span class="p">,</span> <span class="n">backup_url</span><span class="o">=</span><span class="s">'https://www.dropbox.com/s/qj1jlm9w10wmt0u/pancreas.h5ad?dl=1'</span><span class="p">)</span>
<span class="n">pancreas_all</span>
</code></pre></div></div>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>AnnData object with n_obs × n_vars = 14693 × 2448
    obs: 'celltype', 'sample', 'n_genes', 'batch', 'n_counts', 'louvain'
    var: 'n_cells-0', 'n_cells-1', 'n_cells-2', 'n_cells-3'
    uns: 'celltype_colors', 'louvain', 'neighbors', 'pca', 'sample_colors'
    obsm: 'X_pca', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'
</code></pre></div></div>

<p>To be honest, I don’t know what to do about the NaNs in the genes below. If anyone has a good idea for this, I’d love to hear it.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">pancreas_all</span><span class="p">.</span><span class="n">var</span> <span class="c1"># To demonstrate that not every batch recorded the same genes.
</span></code></pre></div></div>

<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n_cells-0</th>
      <th>n_cells-1</th>
      <th>n_cells-2</th>
      <th>n_cells-3</th>
    </tr>
    <tr>
      <th>index</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>A2M</th>
      <td>273.0</td>
      <td>120.0</td>
      <td>262.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>ABAT</th>
      <td>841.0</td>
      <td>988.0</td>
      <td>1052.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>ABCA1</th>
      <td>418.0</td>
      <td>623.0</td>
      <td>468.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>ABCA17P</th>
      <td>63.0</td>
      <td>NaN</td>
      <td>104.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>ABCA7</th>
      <td>410.0</td>
      <td>74.0</td>
      <td>1096.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>MIR663A</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>492.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>LOC100379224</th>
      <td>NaN</td>
      <td>65.0</td>
      <td>125.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>LOC100130093</th>
      <td>NaN</td>
      <td>228.0</td>
      <td>764.0</td>
      <td>635.0</td>
    </tr>
    <tr>
      <th>LOC101928303</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>533.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>COPG</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>635.0</td>
    </tr>
  </tbody>
</table>
<p>2448 rows × 4 columns</p>
</div>

<h2 id="data-processing">data processing</h2>

<p>According to Luecken et al., Scanorama (and in general, other integration methods) are improved by subsetting the data on the set of highly variable genes. We use the CellRanger “flavour” provided in Scanpy. Let’s take the top 1000 highly variable genes. The Scanpy team in general recommends anywhere between 1000 and 5000 HVGs, so you can play with this.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">target_genes</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">sc</span><span class="p">.</span><span class="n">pp</span><span class="p">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">flavor</span><span class="o">=</span><span class="s">'cell_ranger'</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="n">target_genes</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s">'batch'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm
/usr/local/lib/python3.6/dist-packages/scanpy/preprocessing/_highly_variable_genes.py:504: FutureWarning: Slicing a positional slice with .loc is not supported, and will raise TypeError in a future version.  Use .loc with labels or .iloc with positions instead.
  df.loc[:n_top_genes, 'highly_variable'] = True
</code></pre></div></div>

<p>We check the number of HVGs that are highly variable across all $N$ batches. If we do not have enough, we will iterate, taking HVGs from the next \(N-1\), $N - 2, \dots$ batches, until we have the desired number, as specified by <code class="language-plaintext highlighter-rouge">target_genes</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># As we don't have enough target genes, we need to consider the 'next best' HVGs
</span><span class="n">n_batches</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'batch'</span><span class="p">].</span><span class="n">cat</span><span class="p">.</span><span class="n">categories</span><span class="p">)</span>
<span class="c1"># These are the genes that are variable across all batches
</span><span class="n">nbatch1_dispersions</span> <span class="o">=</span> <span class="n">pancreas_all</span><span class="p">.</span><span class="n">var</span><span class="p">[</span><span class="s">'dispersions_norm'</span><span class="p">][</span><span class="n">pancreas_all</span><span class="p">.</span><span class="n">var</span><span class="p">.</span><span class="n">highly_variable_nbatches</span> <span class="o">&gt;</span> <span class="n">n_batches</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">nbatch1_dispersions</span><span class="p">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nbatch1_dispersions</span><span class="p">))</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>176
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Fill up the genes now, using this method from the Theis lab
</span><span class="n">enough</span> <span class="o">=</span> <span class="bp">False</span>
<span class="n">hvg</span> <span class="o">=</span> <span class="n">nbatch1_dispersions</span><span class="p">.</span><span class="n">index</span><span class="p">[:]</span>
<span class="n">not_n_batches</span> <span class="o">=</span> <span class="mi">1</span>

<span class="c1"># We'll go down one by one, until we're selecting HVGs from just a single batch.
</span><span class="k">while</span> <span class="ow">not</span> <span class="n">enough</span><span class="p">:</span>

    <span class="n">target_genes_diff</span> <span class="o">=</span> <span class="n">target_genes</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">hvg</span><span class="p">)</span> <span class="c1"># Get the number of genes we still need to fill up
</span>
    <span class="n">tmp_dispersions</span> <span class="o">=</span> <span class="n">pancreas_all</span><span class="p">.</span><span class="n">var</span><span class="p">[</span><span class="s">'dispersions_norm'</span><span class="p">][</span><span class="n">pancreas_all</span><span class="p">.</span><span class="n">var</span><span class="p">.</span><span class="n">highly_variable_nbatches</span> <span class="o">==</span> <span class="p">(</span><span class="n">n_batches</span> <span class="o">-</span> <span class="n">not_n_batches</span><span class="p">)]</span>

    <span class="c1"># If we haven't hit the target gene numbers, add this to the list and we repeat this iteration
</span>    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tmp_dispersions</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">target_genes_diff</span><span class="p">:</span>

        <span class="n">hvg</span> <span class="o">=</span> <span class="n">hvg</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_dispersions</span><span class="p">.</span><span class="n">index</span><span class="p">)</span>
        <span class="n">not_n_batches</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">else</span><span class="p">:</span>

        <span class="n">tmp_dispersions</span><span class="p">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="n">hvg</span> <span class="o">=</span> <span class="n">hvg</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_dispersions</span><span class="p">.</span><span class="n">index</span><span class="p">[:</span><span class="n">target_genes_diff</span><span class="p">])</span>
        <span class="n">enough</span> <span class="o">=</span> <span class="bp">True</span>
</code></pre></div></div>

<p>We can now subset the data on the highly variable genes.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">pancreas_all</span> <span class="o">=</span> <span class="n">pancreas_all</span><span class="p">[:,</span> <span class="n">hvg</span><span class="p">]</span> <span class="c1"># Filter out genes that do not vary much across cells
</span></code></pre></div></div>

<p>This step does not actually need to be done, as you can see in <code class="language-plaintext highlighter-rouge">pancreas_all.obsm</code> that the PCA and UMAP have already been calculated. But I’ll show it for completneess</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="p">.</span><span class="n">tl</span><span class="p">.</span><span class="n">pca</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s">'arpack'</span><span class="p">)</span> <span class="c1"># Calculate the PCA embeddings
</span><span class="n">sc</span><span class="p">.</span><span class="n">pp</span><span class="p">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">)</span> <span class="c1"># Determine the kNN graph
</span><span class="n">sc</span><span class="p">.</span><span class="n">tl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">)</span> <span class="c1"># Calculate the UMAP
</span></code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="p">.</span><span class="n">pl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s">'batch'</span><span class="p">,</span> <span class="s">'celltype'</span><span class="p">])</span> <span class="c1"># We plot the cell types as well, since it's been already processed.
</span></code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Trying to set attribute `.uns` of view, copying.
</code></pre></div></div>

<p><img src="/assets/images/scanorama/output_17_1.png" alt="png" /></p>

<p>Scanorama requires us to split the data by the batch/sample, so we will do so now.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">pancreas_split</span>  <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">batch</span> <span class="ow">in</span> <span class="n">pancreas_all</span><span class="p">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'batch'</span><span class="p">].</span><span class="n">unique</span><span class="p">():</span>
    <span class="n">pancreas_split</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">[</span><span class="n">pancreas_all</span><span class="p">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'batch'</span><span class="p">]</span><span class="o">==</span><span class="n">batch</span><span class="p">].</span><span class="n">copy</span><span class="p">())</span>
</code></pre></div></div>

<p>In general, one can choose to scale the data, as Luecken et al. do. I’ve found for my own work that it doesn’t make a difference on integration.</p>

<h2 id="running-scanorama">running Scanorama</h2>

<p>We’re now ready to run Scanorama. Scanorama has two main functions, <code class="language-plaintext highlighter-rouge">correct</code> and <code class="language-plaintext highlighter-rouge">integrate</code>, and their Scanpy equivalents, <code class="language-plaintext highlighter-rouge">correct_scanpy</code> and <code class="language-plaintext highlighter-rouge">integrate_scanpy</code>, respectively. The former method is intended for batch correction, while the latter is intended for data integration. There is a distinction between the terms, as described in this <a href="">paper</a>, but there is probably more overlap than the terms suggest. However, there are a couple of differences between these methods, as described <a href="https://github.com/brianhie/scanorama">here</a>.</p>

<p>For both methods, a lower-dimensional embedding is calculated, called <code class="language-plaintext highlighter-rouge">obms['X_scanorama']</code>, which is to be used instead of PCA. For batch correction, the gene expression matrices are also transformed and returned. If one wants to perform both batch correction AND integration, one sets the option <code class="language-plaintext highlighter-rouge">return_dimred=True</code>.</p>

<p>Another difference is that when running <code class="language-plaintext highlighter-rouge">integrate</code>, one can also obtain a ‘representative subset’ of the data using geometric sketching. For those on the mathematics sides of things, the differences between sampling via geometric sketching vs sampling on the data itself is kind of like the difference between the Riemann and Lebesgue integral, respectively. The idea is that geometric sketching produces a better representation and accounts for rarer cell types. One can then use this sketch as a reference for futher integration/analysis. This is also particularly useful if the full dataset is quite large and one wants to speed up computation. For more details on geometric sketching, please see the following paper by <a href="https://doi.org/10.1016/j.cels.2019.05.003">Hie et al. (2020)</a>—it’s one of the only papers in a Cell journal that I’ve seen contain a proof!</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Now we run Scanorama on the split data.
</span><span class="n">corrected</span> <span class="o">=</span> <span class="n">scrama</span><span class="p">.</span><span class="n">correct_scanpy</span><span class="p">(</span><span class="n">pancreas_split</span><span class="p">,</span> <span class="n">return_dimred</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="c1"># Merge the corrected datasets
</span><span class="n">pancreas_all_corrected</span> <span class="o">=</span> <span class="n">corrected</span><span class="p">[</span><span class="mi">0</span><span class="p">].</span><span class="n">concatenate</span><span class="p">(</span><span class="n">corrected</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="n">pancreas_all_corrected</span><span class="p">.</span><span class="n">obs_names_make_unique</span><span class="p">(</span><span class="n">join</span><span class="o">=</span><span class="s">'_'</span><span class="p">)</span>
</code></pre></div></div>

<p>To remind ourselves, let’s look at how well integrated the data are before Scanorama.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Before scanorama
</span><span class="n">sc</span><span class="p">.</span><span class="n">tl</span><span class="p">.</span><span class="n">pca</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s">'arpack'</span><span class="p">)</span>
<span class="n">sc</span><span class="p">.</span><span class="n">pp</span><span class="p">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">)</span>
<span class="n">sc</span><span class="p">.</span><span class="n">tl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="p">.</span><span class="n">pl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s">'batch'</span><span class="p">,</span> <span class="s">'celltype'</span><span class="p">])</span>
</code></pre></div></div>

<p><img src="/assets/images/scanorama/output_26_0.png" alt="png" /></p>

<p>We now use Scanorama’s embedding instead of PCA.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Now rerun this with Scanoramaa
</span><span class="n">sc</span><span class="p">.</span><span class="n">pp</span><span class="p">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">pancreas_all_corrected</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s">'X_scanorama'</span><span class="p">)</span>
<span class="n">sc</span><span class="p">.</span><span class="n">tl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all_corrected</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="p">.</span><span class="n">pl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all_corrected</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s">'batch'</span><span class="p">,</span> <span class="s">'celltype'</span><span class="p">])</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>... storing 'celltype' as categorical
... storing 'sample' as categorical
... storing 'louvain' as categorical
</code></pre></div></div>

<p><img src="/assets/images/scanorama/output_29_1.png" alt="png" /></p>

<p>Ta-da! The batches are better integrated with Scanorama. For comparison, let’s also run BBKNN on this (I actually suspect BBKNN will do better).</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="p">.</span><span class="n">external</span><span class="p">.</span><span class="n">pp</span><span class="p">.</span><span class="n">bbknn</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">batch_key</span><span class="o">=</span><span class="s">'batch'</span><span class="p">)</span> <span class="c1"># This replaces sc.pp.neighbours and is run on PCA.
</span><span class="n">sc</span><span class="p">.</span><span class="n">tl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="p">.</span><span class="n">pl</span><span class="p">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pancreas_all</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s">'batch'</span><span class="p">,</span> <span class="s">'celltype'</span><span class="p">])</span>
</code></pre></div></div>

<p><img src="/assets/images/scanorama/output_32_0.png" alt="png" /></p>


    </div>

</article>
<div class="post-nav"><span></span><a class="next" href="/pseudotime/2021/02/09/scvelo-tutorial.html" title="scVelo">scVelo</a></div><div class="post-related">
      <div>Related Articles</div>
      <ul>
        <li><a class="post-link" href="/hpc/2021/04/07/CondaEnvOnHPC3.html" title="scVelo">Python on HPC3</a></li><li><a class="post-link" href="/pseudotime/2021/02/09/scvelo-tutorial.html" title="scVelo">scVelo</a></li><li><a class="post-link" href="/splicing/2022/01/27/swan.html" title="scVelo">Swan</a></li><li><a class="post-link" href="/long-read/2021/04/19/long-reads.html" title="scVelo">Analysis of long-read transcriptomes</a></li></ul>
    </div><div class="post-comments">  <div id="disqus_thread"></div>
  <script>
    var disqus_config = function () {
      this.page.url = '/integration/2020/12/03/scanorama_demo_pancreas.html';
      this.page.identifier = '/integration/2020/12/03/scanorama_demo_pancreas.html';
    };

    (function() {
      var d = document, s = d.createElement('script');

      s.src = 'https://https-uci-genpals-github-io.disqus.com/embed.js';

      s.setAttribute('data-timestamp', +new Date());
      (d.head || d.body).appendChild(s);
    })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>
</div></section>
</div>


  </section>
  <section class="sidebar" style="margin-left: 15px;">
    <!-- Get sidebar items --></section>
</div>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">
    <div class="site-footer-inner">
      <div>Copyright © 2020-2021 @UCI genPALS</div>
      <div>Powered by <a title="Jekyll is a simple, blog-aware, static site
      generator." href="http://jekyllrb.com/">Jekyll</a> &amp; <a title="Yat, yet
      another theme." href="https://github.com/jeffreytse/jekyll-theme-yat">Yat Theme</a>.</div>
      <div class="footer-col rss-subscribe">Subscribe <a href="/feed.xml">via RSS</a></div>
    </div>
  </div>
</footer>
</body>
</html>
