<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="google-translate-customization" content="108d9124921d80c3-80e20d618ff053c8-g4f02ec6f3dba68b7-c"></meta><!-- Begin Jekyll SEO tag v2.6.1 -->
<title>Analysis of long-read transcriptomes | UCI genPALS</title>
<meta name="generator" content="Jekyll v3.9.0" />
<meta property="og:title" content="Analysis of long-read transcriptomes" />
<meta name="author" content="Fairlie Reese" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Author: Fairlie Reese" />
<meta property="og:description" content="Author: Fairlie Reese" />
<link rel="canonical" href="/long-read/2021/04/19/long-reads.html" />
<meta property="og:url" content="/long-read/2021/04/19/long-reads.html" />
<meta property="og:site_name" content="UCI genPALS" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2021-04-19T00:00:00-07:00" />
<script type="application/ld+json">
{"headline":"Analysis of long-read transcriptomes","url":"/long-read/2021/04/19/long-reads.html","dateModified":"2021-04-19T00:00:00-07:00","datePublished":"2021-04-19T00:00:00-07:00","mainEntityOfPage":{"@type":"WebPage","@id":"/long-read/2021/04/19/long-reads.html"},"author":{"@type":"Person","name":"Fairlie Reese"},"description":"Author: Fairlie Reese","@type":"BlogPosting","@context":"https://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="shortcut icon" href="">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/typeface-noto-sans@0.0.72/index.min.css">
  <link rel="stylesheet"
        href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/default.min.css">
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js"></script>
  <!-- and it's easy to individually load additional languages -->
  <script charset="UTF-8"
          src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/languages/go.min.js"></script>
  <link rel="stylesheet" href="/assets/css/main.css">
  <script src="/assets/js/main.js"></script><link type="application/atom+xml" rel="alternate" href="/feed.xml" title="UCI genPALS" /></head>
<body>



















<header class="site-header site-header-transparent" role="banner">

  <div class="wrapper">
    <div class="site-header-inner"><span class="site-brand"><a class="site-brand-inner" rel="author" href="/">
  <img class="site-favicon" title="UCI genPALS" src="" onerror="this.style.display='none'">
  UCI genPALS
</a>
</span><nav class="site-nav">
          <input type="checkbox" id="nav-trigger" class="nav-trigger" />
          <label for="nav-trigger">
            <span class="menu-icon">
              <svg viewBox="0 0 18 15" width="18px" height="15px">
                <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
              </svg>
            </span>
          </label>

          <div class="trigger"><a class="page-link" href="/about.html">ABOUT</a><a class="page-link" href="/archives.html">ARCHIVES</a><a class="page-link" href="/categories.html">CATEGORIES</a><a class="page-link" href="/events.html">EVENTS</a><a class="page-link" href="/">HOME</a><a class="page-link" href="/resources.html">RESOURCES</a><a class="page-link" href="/symposium.html">GENPALS SYMPOSIUM 2021</a><a class="page-link" href="/tags.html">TAGS</a>




<span class="page-link">

<div id="google_translate_element" style="display: none;">
</div>

<span class="ct-language">
  <ul class="list-unstyled ct-language-dropdown">
    
      <li>
        <a href="#" class="lang-select" data-lang="en">
          
          <img src="https://www.countryflags.io/us/flat/64.png" title="English">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="fr">
          
          <img src="https://www.countryflags.io/fr/flat/64.png" title="Franch">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="zh-CN">
          
          <img src="https://www.countryflags.io/cn/flat/64.png" title="Chinese(Simple)">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="ja">
          
          <img src="https://www.countryflags.io/jp/flat/64.png" title="Japan">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="ko">
          
          <img src="https://www.countryflags.io/kr/flat/64.png" title="Korean">
          
        </a>
      </li>
    
      <li>
        <a href="#" class="lang-select" data-lang="ru">
          
          <img src="https://www.countryflags.io/ru/flat/64.png" title="Russia">
          
        </a>
      </li>
    
  </ul>
</span>

<script type="text/javascript">
function googleTranslateElementInit() {
  new google.translate.TranslateElement({
    pageLanguage: '',
    autoDisplay: false,
    layout: google.translate.TranslateElement.InlineLayout.VERTICAL
  }, 'google_translate_element');

  function restoreLang() {
    var iframe = document.getElementsByClassName('goog-te-banner-frame')[0];
    if (!iframe) return;

    var innerDoc = iframe.contentDocument || iframe.contentWindow.document;
    var restore_el = innerDoc.getElementsByTagName("button");

    for (var i = 0; i < restore_el.length; i++) {
      if (restore_el[i].id.indexOf("restore") >= 0) {
        restore_el[i].click();
        var close_el = innerDoc.getElementsByClassName("goog-close-link");
        close_el[0].click();
        return;
      }
    }
  }

  function triggerHtmlEvent(element, eventName) {
    var event;
    if (document.createEvent) {
      event = document.createEvent('HTMLEvents');
      event.initEvent(eventName, true, true);
      element.dispatchEvent(event);
    } else {
      event = document.createEventObject();
      event.eventType = eventName;
      element.fireEvent('on' + event.eventType, event);
    }
  }

  var googleCombo = document.querySelector("select.goog-te-combo");
  var langSelect = document.querySelector('.ct-language');
  langSelect.addEventListener('click', function(event) {
    if (!event.target) {
      return;
    }

    var selected = document.querySelector('.ct-language .ct-language-selected');
    if (selected) {
      selected.classList.remove('ct-language-selected');
    }

    var target = event.target;
    while (target && target !== langSelect ) {
      if (target.matches('.lang-select')) {
        break;
      }
      target = target.parentElement;
    }

    if (target && target.matches('.lang-select')) {
      var lang = target.getAttribute('data-lang');
      if (googleCombo.value == lang) {
        restoreLang();
      } else {
        target.parentElement.classList.add('ct-language-selected');
        googleCombo.value = lang;
        triggerHtmlEvent(googleCombo, 'change');
      }
    }

    event.preventDefault();
  });
}
</script>

<script type="text/javascript" src="//translate.google.com/translate_a/element.js?cb=googleTranslateElementInit"></script>
</span></div>
        </nav></div>
  </div>
</header>

<script>
  (function() {
    var lastScrollY = getScrollPos().y;
    var documentElement = document.documentElement;

    function storeScrollData() {
      var y = getScrollPos().y;documentElement.setAttribute("data-header-transparent", "");var scrollStatus = "";

      if (y <= 0) {
        scrollStatus = "top";
      } else if ((window.innerHeight + y) >= document.body.offsetHeight) {
        scrollStatus = "bottom";
      } else {
        var isScrollDown = (y - lastScrollY > 0) ? true : false;
        scrollStatus = isScrollDown ? "down" : "up";
      }

      lastScrollY = y;
      documentElement.setAttribute("data-scroll-status", scrollStatus);
    }

    window.addEventListener('scroll', function(e) {
      storeScrollData();
    });

    storeScrollData();
  })();
</script>











































  <section class="page-banner">
    <div class="page-banner-img">
      <div style="background-image: url(/assets/images/banners/ginkgo.jpeg)"></div>
    </div>
    <div class="wrapper">
      <div class="page-banner-inner"><header class="post-header">
  <h1 class="post-title p-name" itemprop="name headline">Analysis of long-read transcriptomes</h1>
  <h3 class="post-subtitle">Application of long-read tools to differentiating PGP1 cells</h3>

  <p class="post-meta">
    <time class="dt-published" datetime="2021-04-19T00:00:00-07:00" itemprop="datePublished"><i class="fa fa-calendar"></i> Apr 19, 2021
    </time>

    





















    <span class="post-reading-time left-vsplit"><i class="fa fa-clock-o"></i> About 1 hour 11 mins</span>
  </p><div class="post-tags"><a class="post-tag" href="/tags.html#long-read">#long-read</a><a class="post-tag" href="/tags.html#python">#python</a><a class="post-tag" href="/tags.html#RNA">#RNA</a><a class="post-tag" href="/tags.html#genome-browser">#genome-browser</a><a class="post-tag" href="/tags.html#splicing">#splicing</a></div></header>
</div>
    </div>
  </section><script>
  function hashLocate(hashValue) {
    hashValue = hashValue.replace(/^.*#h-/, '');
    var element = document.getElementById(hashValue);

    if (!element) {
      return;
    }

    var header = document.querySelector('header.site-header');
    var headerRect = header.getBoundingClientRect();
    var headerTop = Math.floor(headerRect.top);
    var headerHeight = Math.floor(headerRect.height);
    var scrollPos = getScrollPos();
    var offsetY = element.offsetTop - (headerTop + headerHeight + 20);

    if (offsetY == scrollPos.y) {
      return;
    }

    if (headerTop == 0  && offsetY > scrollPos.y) {
      offsetY += headerHeight + 2;
    } else if (headerTop < 0  && offsetY < scrollPos.y) {
      offsetY -= headerHeight - 2;
    }

    smoothScrollTo(offsetY);
  }

  // The first event occurred
  window.addEventListener('load', function(event) {
    if (window.location.hash) {
      hashLocate(window.location.hash);
    }
  });

  // The first event occurred
  window.addEventListener('click', function(event) {
    if (event.target.tagName.toLowerCase() == 'a') {
      hashLocate(event.target.getAttribute('href'));
    }
  });
</script>
<div class="theme-toggle">
  <input type="checkbox" id="theme-switch">
  <label for="theme-switch">
    <div class="toggle"></div>
    <div class="names">
      <p class="light">Light</p>
      <p class="dark">Dark</p>
    </div>
  </label>
</div>




<script>
  (function() {
    var sw = document.getElementById('theme-switch');
    var html = document.getElementsByTagName('html')[0];
    var nightModeOption = ('on' || 'auto').toLowerCase();
    var storage = nightModeOption === 'manual'
        ? localStorage
        : sessionStorage;
    var themeData = loadThemeData();

    function saveThemeData(data) {
      storage.setItem('theme', JSON.stringify(data));
    }

    function loadThemeData() {
      var data = storage.getItem('theme');
      try {
        data = JSON.parse(data ? data : '');
      } catch(e) {
        data = { nightShift: undefined, autoToggleAt: 0 };
        saveThemeData(data);
      }
      return data;
    }

    function handleThemeToggle(nightShift) {
      themeData.nightShift = nightShift;
      saveThemeData(themeData);
      html.dataset.theme = nightShift ? 'dark' : 'light';
      setTimeout(function() {
        sw.checked = nightShift ? true : false;
      }, 50);
    }

    function autoThemeToggle() {
      // Next time point of theme toggle
      var now = new Date();
      var toggleAt = new Date();
      var hours = now.getHours();
      var nightShift = hours >= 19 || hours <=7;

      if (nightShift) {
        if (hours > 7) {
          toggleAt.setDate(toggleAt.getDate() + 1);
        }
        toggleAt.setHours(7);
      } else {
        toggleAt.setHours(19);
      }

      toggleAt.setMinutes(0);
      toggleAt.setSeconds(0);
      toggleAt.setMilliseconds(0)

      var delay = toggleAt.getTime() - now.getTime();

      // auto toggle theme mode
      setTimeout(function() {
        handleThemeToggle(!nightShift);
      }, delay);

      return {
        nightShift: nightShift,
        toggleAt: toggleAt.getTime()
      };
    }

    // Listen the theme toggle event
    sw.addEventListener('change', function(event) {
      handleThemeToggle(event.target.checked);
    });

    if (nightModeOption == 'auto') {
      var data = autoThemeToggle();

      // Toggle theme by local setting
      if (data.toggleAt > themeData.autoToggleAt) {
        themeData.autoToggleAt = data.toggleAt;
        handleThemeToggle(data.nightShift);
      } else {
        handleThemeToggle(themeData.nightShift);
      }
    } else if (nightModeOption == 'manual') {
      handleThemeToggle(themeData.nightShift);
    } else {
      var nightShift = themeData.nightShift;
      if (nightShift === undefined) {
        nightShift = nightModeOption === 'on';
      }
      handleThemeToggle(nightShift);
    }
  })();
</script>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <div class="framework">
  <section class="main">

     <div class="post">
  <section>









<article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

    <div class="post-content e-content" itemprop="articleBody">

      <p>Author: Fairlie Reese</p>

<h2 id="data-download-and-formatting">Data download and formatting</h2>

<h3 id="long-read-data">Long-read data</h3>

<p>Note:</p>
<ul>
  <li>Astrocyte data currently not publically available. This page will be updated when it is!</li>
  <li>These files already contain full-length non-chimeric reads. They have been processed as per the pipeline outlined <a href="https://www.encodeproject.org/documents/7ec9d66a-3b7e-4183-8677-e1df14770b44/@@download/attachment/ENCODE%20Long%20Read%20RNA-Seq%20Analysis%20Pipeline%20%28Human%29.pdf">here</a></li>
</ul>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># download files</span>
wget http://crick.bio.uci.edu/freese/210413_pgp1_hub/encode_files_dl.txt <span class="nb">.</span>
xargs <span class="nt">-L</span> 1 curl <span class="nt">-O</span> <span class="nt">-J</span> <span class="nt">-L</span> &lt; files.txt

<span class="c"># unzip</span>
<span class="nb">gunzip</span> <span class="k">*</span>fastq.gz

<span class="c"># rename them to make sense</span>
<span class="nb">mv </span>ENCFF954UFG.fastq <span class="o">&gt;</span> pgp1_1.fastq
<span class="nb">mv </span>ENCFF251CBB.fastq <span class="o">&gt;</span> pgp1_2.fastq
<span class="nb">mv </span>ENCFF919JFJ.fastq <span class="o">&gt;</span> excite_neuron_1.fastq
<span class="nb">mv </span>ENCFF982WKN.fastq <span class="o">&gt;</span> excite_neuron_2.fastq
<span class="c"># mv ____.fastq &gt; astro_1.fastq</span>
<span class="c"># mv ____.fastq &gt; astro_2.fastq</span>
</code></pre></div></div>

<h3 id="reference-data">Reference data</h3>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># human reference genome</span>
wget https://www.encodeproject.org/files/GRCh38_no_alt_analysis_set_GCA_000001405.15/@@download/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz
<span class="nb">gunzip </span>GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz
<span class="nb">mv </span>GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta hg38.fasta

<span class="c"># human reference genome chrom sizes</span>
wget https://hgdownload-test.gi.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes <span class="nb">.</span>

<span class="c"># gencode v29 human reference transcriptome</span>
wget https://www.encodeproject.org/files/gencode.v29.primary_assembly.annotation_UCSC_names/@@download/gencode.v29.primary_assembly.annotation_UCSC_names.gtf.gz
<span class="nb">gunzip </span>gencode.v29.primary_assembly.annotation_UCSC_names.gtf.gz
<span class="nb">mv </span>gencode.v29.primary_assembly.annotation_UCSC_names.gtf gencode.v29.annotation.gtf
</code></pre></div></div>

<h3 id="set-up-a-small-file-holding-the-names-of-each-sample">Set up a small file holding the names of each sample</h3>
<p>Trust me it’s easier this way</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">touch</span> <span class="s2">"samples.txt"</span>
<span class="nb">printf</span> <span class="s2">"pgp1_1</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> samples.txt
<span class="nb">printf</span> <span class="s2">"pgp1_2</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> samples.txt
<span class="nb">printf</span> <span class="s2">"excite_neuron_1</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> samples.txt
<span class="nb">printf</span> <span class="s2">"excite_neuron_2</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> samples.txt
<span class="nb">printf</span> <span class="s2">"astro_1</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> samples.txt
<span class="nb">printf</span> <span class="s2">"astro_2</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> samples.txt
</code></pre></div></div>

<h2 id="minimap">Minimap</h2>

<p>Map each set of reads to the reference genome. Meant to be run on the cluster.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>module load samtools
module load minimap2/2.17

<span class="nv">genome</span><span class="o">=</span>hg38.fasta

<span class="c"># this will loop through and map each sample in samples.txt!</span>
<span class="k">while </span><span class="nb">read </span>p
<span class="k">do
    </span><span class="nv">fastq</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>.fastq
    <span class="nv">sam</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_mapped.sam
    <span class="nv">log</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_minimap.log
    minimap2 <span class="se">\</span>
        <span class="nt">-t</span> 16 <span class="se">\</span>
        <span class="nt">-ax</span> splice:hq <span class="se">\</span>
        <span class="nt">-uf</span> <span class="se">\</span>
        <span class="nt">--MD</span> <span class="se">\</span>
        <span class="nv">$genome</span> <span class="se">\</span>
        <span class="nv">$fastq</span> <span class="o">&gt;</span> <span class="se">\</span>
        <span class="nv">$sam</span> 2&gt; <span class="se">\</span>
        <span class="nv">$log</span>
<span class="k">done</span> &lt; samples.txt
</code></pre></div></div>

<h2 id="transcriptclean">TranscriptClean</h2>

<p>Correct common long-read sequencing artifacts. Meant to be run on the cluster.</p>

<p>Download TranscriptClean from here: https://github.com/mortazavilab/TranscriptClean, and use the path where you downloaded it to as <code class="language-plaintext highlighter-rouge">tc_path</code> in the following block.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>module load samtools
<span class="nv">tc_path</span><span class="o">=</span>/dfs6/pub/freese/mortazavi_lab/bin/TranscriptClean/
<span class="nv">genome</span><span class="o">=</span>hg38.fasta

<span class="k">while </span><span class="nb">read </span>p
<span class="k">do
    </span><span class="nv">sam</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_mapped.sam
    <span class="nv">bam</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_mapped.bam
    <span class="nv">sort_bam</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_sorted.bam
    <span class="nv">sort_sam</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_sorted.sam

    <span class="c"># first sort the sam file</span>
    samtools view <span class="nt">-Sb</span> <span class="nv">$sam</span> <span class="o">&gt;</span> <span class="nv">$bam</span>
    samtools <span class="nb">sort</span> <span class="nv">$bam</span> <span class="o">&gt;</span> <span class="nv">$sort_bam</span>
    samtols view <span class="nt">-h</span> <span class="nv">$sort_bam</span> <span class="o">&gt;</span> <span class="nv">$sort_sam</span>

    <span class="c"># run TranscriptClean</span>
    python <span class="k">${</span><span class="nv">tc_path</span><span class="k">}</span>/TranscriptClean.py <span class="se">\</span>
        <span class="nt">--sam</span> <span class="nv">$sort_sam</span> <span class="se">\</span>
        <span class="nt">--genome</span> <span class="nv">$genome</span> <span class="se">\</span>
        <span class="nt">-t</span> 16 <span class="se">\</span>
        <span class="nt">--canonOnly</span> <span class="se">\</span>
        <span class="nt">--deleteTmp</span> <span class="se">\</span>
        <span class="nt">--outprefix</span> <span class="nv">$p</span>
<span class="k">done</span> &lt; samples.txt
</code></pre></div></div>

<h2 id="talon">TALON</h2>

<h3 id="talon-label-reads">TALON label reads</h3>

<p>Before running TALON, we have to determine what the nucleotide composition of the end of each read is, which will help us filter for internal priming.</p>

<p>Download and install TALON according to the instructions on the TALON repository: https://github.com/mortazavilab/TALON</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">genome</span><span class="o">=</span>hg38.fasta

<span class="k">while </span><span class="nb">read </span>p
<span class="k">do
    </span><span class="nv">sam</span><span class="o">=</span><span class="k">${</span><span class="nv">p</span><span class="k">}</span>_clean.sam
    talon_label_reads <span class="se">\</span>
        <span class="nt">--f</span> <span class="nv">$sam</span> <span class="se">\</span>
        <span class="nt">--g</span> <span class="nv">$genome</span> <span class="se">\</span>
        <span class="nt">--t</span> 16 <span class="se">\</span>
        <span class="nt">--ar</span> 20  <span class="se">\</span>
        <span class="nt">--deleteTmp</span>  <span class="se">\</span>
        <span class="nt">--o</span> <span class="nv">$p</span>
<span class="k">done</span> &lt; samples.txt
</code></pre></div></div>

<h3 id="talon-database-initialization">TALON database initialization</h3>

<p>Before we run TALON on our reads, we have to add a reference annotation to compare it to</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">annot</span><span class="o">=</span>gencode.v29.annotation.gtf
talon_initialize_database <span class="se">\</span>
    <span class="nt">--f</span> <span class="nv">$annot</span> <span class="se">\</span>
    <span class="nt">--g</span> hg38 <span class="se">\</span>
    <span class="nt">--a</span> gencode_v29 <span class="se">\</span>
    <span class="nt">--l</span> 0 <span class="se">\</span>
    <span class="nt">--idprefix</span> ENCODEH <span class="se">\</span>
    <span class="nt">--5p</span> 500 <span class="se">\</span>
    <span class="nt">--3p</span> 300 <span class="se">\</span>
    <span class="nt">--o</span> pgp1
</code></pre></div></div>

<h3 id="create-a-talon-config-file">Create a TALON config file</h3>

<p>Create a comma-separated file that provides the sample name, sample description, platform, and location of each input sam file.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">touch </span>talon_config.csv
<span class="nb">printf</span> <span class="s2">"pgp1_1,pgp1,SequelI,pgp1_1_labeled.sam</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> talon_config.csv
<span class="nb">printf</span> <span class="s2">"pgp1_2,pgp1,SequelI,pgp1_2_labeled.sam</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> talon_config.csv
<span class="nb">printf</span> <span class="s2">"excite_neuron_1,excitatory_neuron,SequelI,excite_neuron_1_labeled.sam</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> talon_config.csv
<span class="nb">printf</span> <span class="s2">"excite_neuron_2,excitatory_neuron,SequelI,excite_neuron_2_labeled.sam</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> talon_config.csv
<span class="nb">printf</span> <span class="s2">"astro_1,astrocyte,SequelII,astro_1_labeled.sam</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> talon_config.csv
<span class="nb">printf</span> <span class="s2">"astro_2,astrocyte,SequelII,astro_2_labeled.sam</span><span class="se">\n</span><span class="s2">"</span> <span class="o">&gt;&gt;</span> talon_config.csv
</code></pre></div></div>

<h3 id="run-talon">Run TALON</h3>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>talon <span class="se">\</span>
    <span class="nt">--f</span> talon_config.csv <span class="se">\</span>
    <span class="nt">--db</span> pgp1.db <span class="se">\</span>
    <span class="nt">--build</span> hg38 <span class="se">\</span>
    <span class="nt">--t</span> 16 <span class="se">\</span>
    <span class="nt">--o</span> <span class="nv">$pgp1</span>
</code></pre></div></div>

<h3 id="filter-output-transcripts">Filter output transcripts</h3>

<p>Filter novel transcripts for internal priming and for reproducibility</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">db</span><span class="o">=</span>pgp1.db
talon_filter_transcripts <span class="se">\</span>
    <span class="nt">--db</span> <span class="nv">$db</span> <span class="se">\</span>
    <span class="nt">-a</span> gencode_v29 <span class="se">\</span>
    <span class="nt">--maxFracA</span><span class="o">=</span>0.5 <span class="se">\</span>
    <span class="nt">--minCount</span><span class="o">=</span>5 <span class="se">\</span>
    <span class="nt">--minDatasets</span><span class="o">=</span>2 <span class="se">\</span>
    <span class="nt">--o</span> pgp1_pass_list.csv
</code></pre></div></div>

<h3 id="create-an-unfiltered-abundance-file">Create an unfiltered abundance file</h3>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">db</span><span class="o">=</span>pgp1.db
talon_abundance <span class="se">\</span>
    <span class="nt">--db</span> <span class="nv">$db</span> <span class="se">\</span>
    <span class="nt">-a</span> gencode_v29 <span class="se">\</span>
    <span class="nt">-b</span> hg38 <span class="se">\</span>
    <span class="nt">--o</span> pgp1
</code></pre></div></div>

<h3 id="visualizing-talon-output">Visualizing TALON output</h3>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="n">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="n">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="n">st</span>
<span class="kn">import</span> <span class="nn">statsmodels.stats</span> <span class="k">as</span> <span class="n">stm</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>
<span class="kn">import</span> <span class="nn">swan_vis</span> <span class="k">as</span> <span class="n">swan</span>
<span class="kn">from</span> <span class="nn">matplotlib.colors</span> <span class="kn">import</span> <span class="n">ListedColormap</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="n">mpl</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">shlex</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">get_talon_nov_colors</span><span class="p">():</span>
    <span class="n">c_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s">'Known'</span><span class="p">:</span> <span class="s">'#009E73'</span><span class="p">,</span>
              <span class="s">'ISM'</span><span class="p">:</span> <span class="s">'#0072B2'</span><span class="p">,</span>
              <span class="s">'NIC'</span><span class="p">:</span> <span class="s">'#D55E00'</span><span class="p">,</span>
              <span class="s">'NNC'</span><span class="p">:</span> <span class="s">'#E69F00'</span><span class="p">,</span>
              <span class="s">'Antisense'</span><span class="p">:</span> <span class="s">'#000000'</span><span class="p">,</span>
              <span class="s">'Intergenic'</span><span class="p">:</span> <span class="s">'#CC79A7'</span><span class="p">,</span>
              <span class="s">'Genomic'</span><span class="p">:</span> <span class="s">'#F0E442'</span><span class="p">}</span>
    <span class="n">order</span> <span class="o">=</span> <span class="p">[</span><span class="s">'Known'</span><span class="p">,</span> <span class="s">'ISM'</span><span class="p">,</span> <span class="s">'NIC'</span><span class="p">,</span> <span class="s">'NNC'</span><span class="p">,</span> <span class="s">'Antisense'</span><span class="p">,</span> <span class="s">'Intergenic'</span><span class="p">,</span> <span class="s">'Genomic'</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span>

<span class="k">def</span> <span class="nf">add_perc</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
    <span class="n">total</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">feature</span><span class="p">].</span><span class="nb">sum</span><span class="p">()</span>
    <span class="n">ylim</span> <span class="o">=</span> <span class="n">ax</span><span class="p">.</span><span class="n">get_ylim</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">ax</span><span class="p">.</span><span class="n">patches</span><span class="p">:</span>
        <span class="n">percentage</span> <span class="o">=</span> <span class="s">'{:.1f}%'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="mi">100</span> <span class="o">*</span> <span class="n">p</span><span class="p">.</span><span class="n">get_height</span><span class="p">()</span><span class="o">/</span><span class="n">total</span><span class="p">)</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">p</span><span class="p">.</span><span class="n">get_x</span><span class="p">()</span> <span class="o">+</span> <span class="n">p</span><span class="p">.</span><span class="n">get_width</span><span class="p">()</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">-</span> <span class="mf">0.4</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">p</span><span class="p">.</span><span class="n">get_y</span><span class="p">()</span> <span class="o">+</span> <span class="n">p</span><span class="p">.</span><span class="n">get_height</span><span class="p">()</span> <span class="o">+</span> <span class="n">ylim</span><span class="o">*</span><span class="mf">0.00625</span>
        <span class="n">ax</span><span class="p">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">percentage</span><span class="p">,</span> <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">),</span> <span class="n">size</span> <span class="o">=</span> <span class="mi">14</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_read_novelty</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">opref</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span>
                      <span class="n">ylim</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                      <span class="n">datasets</span><span class="o">=</span><span class="s">'all'</span><span class="p">):</span>
    <span class="n">sns</span><span class="p">.</span><span class="n">set_context</span><span class="p">(</span><span class="s">"paper"</span><span class="p">,</span> <span class="n">font_scale</span><span class="o">=</span><span class="mf">1.6</span><span class="p">)</span>
    <span class="n">sns</span><span class="p">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s">'white'</span><span class="p">)</span>


    <span class="n">temp</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># filter on datasets
</span>    <span class="k">if</span> <span class="n">datasets</span> <span class="o">!=</span> <span class="s">'all'</span><span class="p">:</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">temp</span><span class="p">.</span><span class="n">dataset</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">datasets</span><span class="p">)]</span>

    <span class="c1"># count number of reads per cat
</span>    <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[[</span><span class="s">'transcript_novelty'</span><span class="p">,</span> <span class="s">'read_name'</span><span class="p">]].</span><span class="n">groupby</span><span class="p">(</span><span class="s">'transcript_novelty'</span><span class="p">).</span><span class="n">count</span><span class="p">()</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'read_name'</span><span class="p">:</span><span class="s">'counts'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>

    <span class="c1"># actual plotting
</span>    <span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="p">.</span><span class="n">catplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">temp</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">'transcript_novelty'</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="s">'counts'</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">'bar'</span><span class="p">,</span>
                <span class="n">palette</span><span class="o">=</span><span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)</span>
    <span class="p">[</span><span class="n">plt</span><span class="p">.</span><span class="n">setp</span><span class="p">(</span><span class="n">ax</span><span class="p">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span> <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">g</span><span class="p">.</span><span class="n">axes</span><span class="p">.</span><span class="n">flat</span><span class="p">]</span>
    <span class="n">g</span><span class="p">.</span><span class="n">set_ylabels</span><span class="p">(</span><span class="s">'Reads'</span><span class="p">)</span>
    <span class="n">g</span><span class="p">.</span><span class="n">set_xlabels</span><span class="p">(</span><span class="s">'Transcript novelty'</span><span class="p">)</span>

    <span class="c1"># add percentage labels
</span>    <span class="n">ax</span> <span class="o">=</span> <span class="n">g</span><span class="p">.</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">add_perc</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">temp</span><span class="p">,</span> <span class="s">'counts'</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ylim</span><span class="p">:</span>
        <span class="n">g</span><span class="p">.</span><span class="nb">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">ylim</span><span class="p">))</span>

<span class="c1">#     # add title
#     if not title:
#         g.fig.suptitle('Reads per novelty category')
#     else:
#         g.fig.suptitle('{} reads per novelty category'.format(title))
</span>
    <span class="c1"># save figure
</span>    <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_read_novelty'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">opref</span><span class="p">)</span>
    <span class="n">g</span><span class="p">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">fname</span><span class="o">+</span><span class="s">'.pdf'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_transcript_novelty</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">oprefix</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span> \
                            <span class="n">ylim</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                            <span class="n">whitelist</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">datasets</span><span class="o">=</span><span class="s">'all'</span><span class="p">,</span> <span class="n">save_type</span><span class="o">=</span><span class="s">'pdf'</span><span class="p">):</span>
    <span class="n">sns</span><span class="p">.</span><span class="n">set_context</span><span class="p">(</span><span class="s">'paper'</span><span class="p">,</span> <span class="n">font_scale</span><span class="o">=</span><span class="mf">1.6</span><span class="p">)</span>
    <span class="n">sns</span><span class="p">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s">'white'</span><span class="p">)</span>

    <span class="n">temp</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># remove transcripts that are not on whitelist
</span>    <span class="k">if</span> <span class="n">whitelist</span><span class="p">:</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">temp</span><span class="p">.</span><span class="n">transcript_ID</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">whitelist</span><span class="p">)]</span>

    <span class="c1"># filter on datasets
</span>    <span class="k">if</span> <span class="n">datasets</span> <span class="o">!=</span> <span class="s">'all'</span><span class="p">:</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">temp</span><span class="p">.</span><span class="n">dataset</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">datasets</span><span class="p">)]</span>

    <span class="c1"># count number of isoforms per cat
</span>    <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[[</span><span class="s">'transcript_ID'</span><span class="p">,</span> <span class="s">'transcript_novelty'</span><span class="p">,</span> <span class="s">'read_name'</span><span class="p">]].</span><span class="n">groupby</span><span class="p">([</span><span class="s">'transcript_ID'</span><span class="p">,</span> <span class="s">'transcript_novelty'</span><span class="p">]).</span><span class="n">count</span><span class="p">()</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'read_name'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="s">'transcript_novelty'</span><span class="p">).</span><span class="n">count</span><span class="p">()</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'transcript_ID'</span><span class="p">:</span> <span class="s">'counts'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">temp</span><span class="p">)</span>

    <span class="c1"># actual plotting
</span>    <span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="p">.</span><span class="n">catplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">temp</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">'transcript_novelty'</span><span class="p">,</span>
                <span class="n">y</span><span class="o">=</span><span class="s">'counts'</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">'bar'</span><span class="p">,</span>
                <span class="n">palette</span><span class="o">=</span><span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)</span>
    <span class="p">[</span><span class="n">plt</span><span class="p">.</span><span class="n">setp</span><span class="p">(</span><span class="n">ax</span><span class="p">.</span><span class="n">get_xticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span> <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">g</span><span class="p">.</span><span class="n">axes</span><span class="p">.</span><span class="n">flat</span><span class="p">]</span>
    <span class="n">g</span><span class="p">.</span><span class="n">set_ylabels</span><span class="p">(</span><span class="s">'Isoforms'</span><span class="p">)</span>
    <span class="n">g</span><span class="p">.</span><span class="n">set_xlabels</span><span class="p">(</span><span class="s">'Transcript novelty'</span><span class="p">)</span>

    <span class="c1"># add percentage labels
</span>    <span class="n">ax</span> <span class="o">=</span> <span class="n">g</span><span class="p">.</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">add_perc</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">temp</span><span class="p">,</span> <span class="s">'counts'</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ylim</span><span class="p">:</span>
        <span class="n">g</span><span class="p">.</span><span class="nb">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">ylim</span><span class="p">))</span>

<span class="c1">#     # add title
#     if not title:
#         g.fig.suptitle('Transcript models per novelty category')
#     else:
#         g.fig.suptitle('{} transcript models per novelty category'.format(title))
</span>
    <span class="c1"># save figure
</span>    <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_isoform_novelty'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">oprefix</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">save_type</span> <span class="o">==</span> <span class="s">'png'</span><span class="p">:</span>
        <span class="n">g</span><span class="p">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">fname</span><span class="o">+</span><span class="s">'.png'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">save_type</span> <span class="o">==</span> <span class="s">'pdf'</span><span class="p">:</span>
        <span class="n">g</span><span class="p">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">fname</span><span class="o">+</span><span class="s">'.pdf'</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>

    <span class="n">plt</span><span class="p">.</span><span class="n">show</span><span class="p">()</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">clf</span><span class="p">()</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># number / proportion of reads per novelty category
</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_talon_read_annot.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span> <span class="o">=</span> <span class="n">get_talon_nov_colors</span><span class="p">()</span>
<span class="n">opref</span> <span class="o">=</span> <span class="s">'pgp1'</span>
<span class="n">plot_read_novelty</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">opref</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span>
                      <span class="n">ylim</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                      <span class="n">datasets</span><span class="o">=</span><span class="s">'all'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  transcript_novelty   counts
0          Antisense   137767
1            Genomic   143842
2                ISM   393877
3         Intergenic    16599
4              Known  3068807
5                NIC   180012
6                NNC   109720
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_8_1.png" alt="png" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># unfiltered transcripts per novelty category
</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_talon_read_annot.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span> <span class="o">=</span> <span class="n">get_talon_nov_colors</span><span class="p">()</span>
<span class="n">datasets</span> <span class="o">=</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">,</span> <span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">,</span> <span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">]</span>
<span class="n">opref</span> <span class="o">=</span> <span class="s">'pgp1'</span>
<span class="n">plot_transcript_novelty</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">opref</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span> \
                    <span class="n">ylim</span><span class="o">=</span><span class="mi">90000</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s">'Unfiltered'</span><span class="p">,</span> <span class="n">datasets</span><span class="o">=</span><span class="s">'all'</span><span class="p">,</span> <span class="n">save_type</span><span class="o">=</span><span class="s">'pdf'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  transcript_novelty  counts
0          Antisense   34773
1            Genomic   33150
2                ISM   83895
3         Intergenic    8044
4              Known   36205
5                NIC   67933
6                NNC   64771
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_9_1.png" alt="png" /></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;Figure size 432x288 with 0 Axes&gt;
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># filtered transcripts per novelty category
</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_talon_read_annot.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span> <span class="o">=</span> <span class="n">get_talon_nov_colors</span><span class="p">()</span>
<span class="n">datasets</span> <span class="o">=</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">,</span> <span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">,</span> <span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">]</span>
<span class="n">opref</span> <span class="o">=</span> <span class="s">'pgp1'</span>
<span class="n">pass_list</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_pass_list.csv'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'gene_ID'</span><span class="p">,</span> <span class="s">'transcript_ID'</span><span class="p">])</span>
<span class="n">pass_list</span> <span class="o">=</span> <span class="n">pass_list</span><span class="p">.</span><span class="n">transcript_ID</span><span class="p">.</span><span class="n">unique</span><span class="p">().</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">plot_transcript_novelty</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">opref</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span><span class="p">,</span> \
                            <span class="n">ylim</span><span class="o">=</span><span class="mi">45000</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s">'Filtered'</span><span class="p">,</span>
                            <span class="n">whitelist</span><span class="o">=</span><span class="n">pass_list</span><span class="p">,</span> <span class="n">datasets</span><span class="o">=</span><span class="s">'all'</span><span class="p">,</span> <span class="n">save_type</span><span class="o">=</span><span class="s">'pdf'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  transcript_novelty  counts
0          Antisense     414
1                ISM    2058
2         Intergenic      32
3              Known   36205
4                NIC    1108
5                NNC     414
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_10_1.png" alt="png" /></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;Figure size 432x288 with 0 Axes&gt;
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_talon_read_annot.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">datasets</span> <span class="o">=</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">,</span> <span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">,</span> <span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">]</span>
<span class="n">opref</span> <span class="o">=</span> <span class="s">'pgp1'</span>
<span class="n">pass_list</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_pass_list.csv'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'gene_ID'</span><span class="p">,</span> <span class="s">'transcript_ID'</span><span class="p">])</span>
<span class="n">pass_list</span> <span class="o">=</span> <span class="n">pass_list</span><span class="p">.</span><span class="n">transcript_ID</span><span class="p">.</span><span class="n">unique</span><span class="p">().</span><span class="n">tolist</span><span class="p">()</span>

<span class="k">print</span><span class="p">(</span><span class="s">'Number of reads before filtering: {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">.</span><span class="n">index</span><span class="p">)))</span>
<span class="k">print</span><span class="p">(</span><span class="s">'Number of reads after filtering: {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="n">transcript_ID</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">pass_list</span><span class="p">)].</span><span class="n">index</span><span class="p">)))</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Number of reads before filtering: 4050624
Number of reads after filtering: 3377394
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Distribution of read lengths
</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'pgp1_talon_read_annot.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">sns</span><span class="p">.</span><span class="n">set_context</span><span class="p">(</span><span class="s">'paper'</span><span class="p">,</span> <span class="n">font_scale</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="p">.</span><span class="n">displot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s">'read_length'</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">'kde'</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">ax</span><span class="p">.</span><span class="nb">set</span><span class="p">(</span><span class="n">xlabel</span><span class="o">=</span><span class="s">'Read length'</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s">'KDE of reads'</span><span class="p">,</span>
      <span class="n">title</span><span class="o">=</span><span class="s">''</span><span class="p">,</span> <span class="n">xlim</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">7500</span><span class="p">),</span>
      <span class="n">xticks</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2500</span><span class="p">,</span> <span class="mi">5000</span><span class="p">,</span> <span class="mi">7500</span><span class="p">])</span>
<span class="n">plt</span><span class="p">.</span><span class="n">savefig</span><span class="p">(</span><span class="s">'{}_read_length_kde.pdf'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">opref</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s">'tight'</span><span class="p">)</span>
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_12_0.png" alt="png" /></p>

<h2 id="tss-calling">TSS calling</h2>

<h3 id="subset-tss-reads">Subset TSS reads</h3>

<p>For TSSs, we’ll consider Known, prefix ISM, NIC, and NNC reads, as these are more likely to contain putative 5’ ends.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">annot</span> <span class="o">=</span> <span class="s">'pgp1_talon_read_annot.tsv'</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">tss_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="n">transcript_novelty</span><span class="p">.</span><span class="n">isin</span><span class="p">([</span><span class="s">'Known'</span><span class="p">,</span> <span class="s">'NIC'</span><span class="p">,</span> <span class="s">'NNC'</span><span class="p">,</span> <span class="s">'ISM'</span><span class="p">])]</span>
<span class="n">tss_df</span> <span class="o">=</span> <span class="n">tss_df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">tss_df</span><span class="p">.</span><span class="n">ISM_subtype</span><span class="p">.</span><span class="n">isin</span><span class="p">([</span><span class="s">'None'</span><span class="p">,</span> <span class="s">'Prefix'</span><span class="p">,</span> <span class="s">'Both'</span><span class="p">])]</span>
<span class="n">tss_reads</span> <span class="o">=</span> <span class="n">tss_df</span><span class="p">.</span><span class="n">read_name</span><span class="p">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># tss
</span><span class="n">fname</span> <span class="o">=</span> <span class="s">'tss_read_names.txt'</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">tss_reads</span><span class="p">:</span>
        <span class="n">ofile</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">r</span><span class="o">+</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">)</span>
</code></pre></div></div>

<p>This part should be run on the cluster or using bash. It relies on the existence of an intermediate merged bam file that TALON makes when it is running.</p>

<p>The <code class="language-plaintext highlighter-rouge">picard_path</code> variable refers to wherever <a href="https://broadinstitute.github.io/picard/">Picard</a> is installed.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>module load samtools
<span class="nv">picard_path</span><span class="o">=</span>/opt/apps/picard-tools/1.87/

<span class="nv">in_bam</span><span class="o">=</span>talon_tmp/merged.bam
<span class="nv">out_bam</span><span class="o">=</span>talon_tmp/merged_rg.bam
<span class="nv">tss_reads</span><span class="o">=</span>tss_read_names.txt
<span class="nv">tss_out</span><span class="o">=</span>tss_reads.bam

<span class="c"># first add the RG header tags for each RG</span>
<span class="nv">rg_tags</span><span class="o">=</span><span class="s2">"pgp1_1 pgp1_2 excite_neuron_1 excite_neuron_2 astro_1 astro_2"</span>
samtools view <span class="nt">-H</span> <span class="nv">$in_bam</span> <span class="o">&gt;</span> header.sam
<span class="k">for </span>tag <span class="k">in</span> <span class="nv">$rg_tags</span>
<span class="k">do
    </span><span class="nb">printf</span> <span class="s1">'@RG\tID:${tag}\tPL:PacBio\tSM:PGP1\n'</span> <span class="o">&gt;&gt;</span> header.sam
<span class="k">done
</span>samtools reheader header.sam <span class="nv">$in_bam</span> <span class="o">&gt;</span> <span class="nv">$out_bam</span>

<span class="c"># then limit to the known, prefix ISM, NIC, and NNC reads</span>
java <span class="nt">-jar</span> <span class="k">${</span><span class="nv">picard_path</span><span class="k">}</span>FilterSamReads.jar <span class="se">\</span>
    <span class="nv">I</span><span class="o">=</span><span class="nv">$out_bam</span> <span class="se">\</span>
    <span class="nv">O</span><span class="o">=</span><span class="nv">$tss_out</span> <span class="se">\</span>
    <span class="nv">READ_LIST_FILE</span><span class="o">=</span><span class="nv">$tss_reads</span> <span class="se">\</span>
    <span class="nv">FILTER</span><span class="o">=</span>includeReadList <span class="se">\</span>
    <span class="nv">CREATE_INDEX</span><span class="o">=</span><span class="nb">true</span>

<span class="c"># index bam file</span>
samtools index <span class="nv">$tss_out</span>
</code></pre></div></div>

<h3 id="call-tsss">Call TSSs</h3>

<p>Call transcription start sites using a TSS caller that, roughly, calls peaks on read starts from long read data.</p>

<p>You can download the tool from here: https://github.com/ENCODE-AWG/tss-annotation. Your <code class="language-plaintext highlighter-rouge">tss_dir</code> should be wherever you install it.</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">tss_dir</span><span class="o">=</span>~/mortazavi_lab/bin/tss-annotation/long_read/
python <span class="k">${</span><span class="nv">tss_dir</span><span class="k">}</span>pacbio_to_tss.py <span class="se">\</span>
    <span class="nt">-i</span> tss_reads.bam <span class="se">\</span>
    <span class="nt">--window-size</span><span class="o">=</span>50 <span class="se">\</span>
    <span class="nt">--expression-threshold</span><span class="o">=</span>2 <span class="se">\</span>
    <span class="nt">-o</span> unfilt_tss.bed <span class="se">\</span>
    <span class="nt">-r</span> <span class="se">\</span>
    <span class="nt">-n</span> rev_tss.bw <span class="se">\</span>
    <span class="nt">-p</span> fwd_tss.bw
</code></pre></div></div>

<p>Get the read names associated with each TSS peak so that we can associate TSSs with genes</p>
<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">merged_bam</span><span class="o">=</span>talon_tmp/merged.bam
python <span class="k">${</span><span class="nv">tss_dir</span><span class="k">}</span>tss_reads.py <span class="se">\</span>
    <span class="nt">-i</span> <span class="nv">$merged</span>.bam <span class="se">\</span>
    <span class="nt">-r</span> unfilt_tss.bed <span class="se">\</span>
    <span class="nt">-o</span> tss_reads.bed
</code></pre></div></div>

<h3 id="filter-tsss-based-on-number-of-reads-and-expression-within-gene">Filter TSSs based on number of reads and expression within gene</h3>

<p>Require TSSs to have at least 2 supporting reads and at least 10 percent of the number of reads that the top-expressed TSS in the gene has.</p>

<p>Additionally save TSS expression in a counts matrix.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># filter a list of TSSs for each gene
</span><span class="n">annot</span> <span class="o">=</span> <span class="s">'pgp1_talon_read_annot.tsv'</span>
<span class="n">tss_reads</span> <span class="o">=</span> <span class="s">'tss_reads.bed'</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>

<span class="c1"># remove sirvs and erccs
</span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="p">.</span><span class="n">chrom</span><span class="p">.</span><span class="nb">str</span><span class="p">.</span><span class="n">contains</span><span class="p">(</span><span class="s">'SIRV'</span><span class="p">)]</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="p">.</span><span class="n">chrom</span><span class="p">.</span><span class="nb">str</span><span class="p">.</span><span class="n">contains</span><span class="p">(</span><span class="s">'ERCC'</span><span class="p">)]</span>

<span class="n">ends</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tss_reads</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
            <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'chrom'</span><span class="p">,</span> <span class="s">'start'</span><span class="p">,</span> <span class="s">'end'</span><span class="p">,</span> <span class="s">'read_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">,</span> <span class="s">'strand'</span><span class="p">])</span>

<span class="c1"># merge on read name
</span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ends</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'inner'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s">'read_name'</span><span class="p">)</span>

<span class="c1"># groupby on gene and tss
</span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s">'read_name'</span><span class="p">,</span> <span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">]].</span><span class="n">groupby</span><span class="p">([</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">]).</span><span class="n">count</span><span class="p">()</span>

<span class="c1"># filter tsss for those that have &gt;10% of the reads
# for the most highly-expressed tss of the gene
</span><span class="n">df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">df</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'read_name'</span><span class="p">:</span><span class="s">'count'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">temp</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="nb">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s">'count'</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="n">annot_gene_name</span><span class="o">==</span><span class="n">x</span><span class="p">.</span><span class="n">annot_gene_name</span><span class="p">,</span> <span class="s">'count'</span><span class="p">].</span><span class="nb">max</span><span class="p">()</span><span class="o">*</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>

<span class="n">temp</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">temp</span><span class="p">.</span><span class="n">annot_gene_name</span> <span class="o">==</span> <span class="s">'MBP'</span><span class="p">]</span>

<span class="c1"># assign each TSS a name, and quantify TSS exp
</span><span class="n">temp</span><span class="p">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s">'annot_gene_name'</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">temp</span><span class="p">[</span><span class="s">'tss_id_2'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">nan</span>
<span class="n">prev_gene</span> <span class="o">=</span> <span class="bp">None</span>
<span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">temp</span><span class="p">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">curr_gene</span> <span class="o">=</span> <span class="n">entry</span><span class="p">.</span><span class="n">annot_gene_name</span>
    <span class="k">if</span> <span class="n">curr_gene</span> <span class="o">!=</span> <span class="n">prev_gene</span><span class="p">:</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="n">prev_gene</span> <span class="o">=</span> <span class="n">curr_gene</span>
    <span class="n">temp</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ind</span><span class="p">,</span> <span class="s">'tss_id_2'</span><span class="p">]</span> <span class="o">=</span> <span class="s">'{}_{}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">curr_gene</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>

<span class="c1"># merged called TSSs back in with read annot
</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">ends</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tss_reads</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
            <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'chrom'</span><span class="p">,</span> <span class="s">'start'</span><span class="p">,</span> <span class="s">'end'</span><span class="p">,</span> <span class="s">'read_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">,</span> <span class="s">'strand'</span><span class="p">])</span>
<span class="n">ends</span> <span class="o">=</span> <span class="n">ends</span><span class="p">[[</span><span class="s">'read_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">]]</span>

<span class="c1"># dump filtered TSSs to bed
</span><span class="n">fname</span> <span class="o">=</span> <span class="s">'unfilt_tss.bed'</span>
<span class="n">end_regions</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
            <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'chrom'</span><span class="p">,</span> <span class="s">'start'</span><span class="p">,</span> <span class="s">'end'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">,</span> <span class="s">'read_count'</span><span class="p">,</span> <span class="s">'strand'</span><span class="p">,</span>
                   <span class="s">'sth1'</span><span class="p">,</span> <span class="s">'sth2'</span><span class="p">,</span> <span class="s">'sth3'</span><span class="p">,</span> <span class="s">'sth4'</span><span class="p">,</span> <span class="s">'sth5'</span><span class="p">])</span>
<span class="n">filt_ends</span> <span class="o">=</span> <span class="n">end_regions</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">temp</span><span class="p">[[</span><span class="s">'tss_id'</span><span class="p">,</span> <span class="s">'tss_id_2'</span><span class="p">]],</span> <span class="n">how</span><span class="o">=</span><span class="s">'inner'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s">'tss_id'</span><span class="p">)</span>
<span class="n">filt_ends</span><span class="p">[</span><span class="s">'score'</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">filt_ends</span> <span class="o">=</span> <span class="n">filt_ends</span><span class="p">[[</span><span class="s">'chrom'</span><span class="p">,</span> <span class="s">'start'</span><span class="p">,</span> <span class="s">'end'</span><span class="p">,</span> <span class="s">'tss_id_2'</span><span class="p">,</span> <span class="s">'score'</span><span class="p">,</span> <span class="s">'strand'</span><span class="p">]]</span>
<span class="n">fname</span> <span class="o">=</span> <span class="s">'filt_tss.bed'</span>
<span class="n">filt_ends</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c1"># merge on read name
</span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ends</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'inner'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s">'read_name'</span><span class="p">)</span>
<span class="n">temp</span> <span class="o">=</span> <span class="n">temp</span><span class="p">[[</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">,</span> <span class="s">'tss_id_2'</span><span class="p">]]</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">temp</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'inner'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">])</span>

<span class="c1"># format like talon ab
</span><span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'annot_gene_id'</span><span class="p">,</span> <span class="s">'dataset'</span><span class="p">,</span> <span class="s">'tss_id_2'</span><span class="p">]</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="o">+</span><span class="p">[</span><span class="s">'read_name'</span><span class="p">]].</span><span class="n">groupby</span><span class="p">(</span><span class="n">cols</span><span class="p">).</span><span class="n">count</span><span class="p">().</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">df</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'read_name'</span><span class="p">:</span><span class="s">'counts'</span><span class="p">,</span> <span class="s">'tss_id_2'</span><span class="p">:</span><span class="s">'tss_id'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'annot_gene_id'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="s">'dataset'</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s">'counts'</span><span class="p">)</span>
<span class="n">df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">rename_axis</span><span class="p">(</span><span class="bp">None</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">df</span><span class="p">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="c1"># i guess sum over genes with the same name smh
</span><span class="n">df</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'annot_gene_id'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">groupby</span><span class="p">([</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">]).</span><span class="nb">sum</span><span class="p">().</span><span class="n">reset_index</span><span class="p">()</span>


<span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">'pgp1_tss_talon_abundance.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:29: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)
</code></pre></div></div>

<h2 id="swan">Swan</h2>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">annot</span> <span class="o">=</span> <span class="s">'gencode.v29.annotation.gtf'</span>

<span class="c1"># create a config file
</span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'samples.txt'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'col'</span><span class="p">])</span>
<span class="n">df</span><span class="p">[</span><span class="s">'fname'</span><span class="p">]</span> <span class="o">=</span> <span class="s">'pgp1.db'</span>
<span class="n">df</span><span class="p">[</span><span class="s">'counts_file'</span><span class="p">]</span> <span class="o">=</span> <span class="s">'pgp1_talon_abundance.tsv'</span>
<span class="n">df</span><span class="p">[</span><span class="s">'count_cols'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s">'col'</span><span class="p">]</span>
<span class="n">df</span><span class="p">[</span><span class="s">'tid_col'</span><span class="p">]</span> <span class="o">=</span> <span class="s">'annot_transcript_id'</span>
<span class="n">df</span><span class="p">[</span><span class="s">'dataset_name'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s">'col'</span><span class="p">]</span>
<span class="n">df</span><span class="p">[</span><span class="s">'whitelist'</span><span class="p">]</span> <span class="o">=</span> <span class="s">'pgp1_pass_list.csv'</span>
<span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">'swan_config.tsv'</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c1"># initialize SwanGraph
</span><span class="n">sg</span> <span class="o">=</span> <span class="n">swan</span><span class="p">.</span><span class="n">SwanGraph</span><span class="p">()</span>

<span class="c1"># add annotation GTF
</span><span class="n">sg</span><span class="p">.</span><span class="n">add_annotation</span><span class="p">(</span><span class="n">annot</span><span class="p">)</span>

<span class="c1"># add datasets and save graph
</span><span class="n">sg</span><span class="p">.</span><span class="n">add_datasets</span><span class="p">(</span><span class="s">'swan_config.tsv'</span><span class="p">)</span>
<span class="n">sg</span><span class="p">.</span><span class="n">save_graph</span><span class="p">(</span><span class="s">'swan'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Adding dataset annotation to the SwanGraph

Adding dataset pgp1_1 to the SwanGraph

Adding dataset pgp1_2 to the SwanGraph

Adding dataset excite_neuron_1 to the SwanGraph

Adding dataset excite_neuron_2 to the SwanGraph

Adding dataset astro_1 to the SwanGraph

Adding dataset astro_2 to the SwanGraph
Saving graph as swan.p
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># detect IR and ES
</span><span class="n">g_es</span><span class="p">,</span> <span class="n">t_es</span><span class="p">,</span> <span class="n">e_es</span> <span class="o">=</span> <span class="n">sg</span><span class="p">.</span><span class="n">find_es_genes</span><span class="p">()</span>
<span class="n">g_ir</span><span class="p">,</span> <span class="n">t_ir</span><span class="p">,</span> <span class="n">e_ir</span> <span class="o">=</span> <span class="n">sg</span><span class="p">.</span><span class="n">find_ir_genes</span><span class="p">()</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Analyzing 866 intronic edges for ES
Found 237 novel es events in 275 transcripts.
Analyzing 2411 exonic edges for IR
Found 71 novel ir events from 82 transcripts.
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># plot gene summary/transcript path graph for a gene with IR (SARS)
</span><span class="n">sg</span><span class="p">.</span><span class="n">plot_graph</span><span class="p">(</span><span class="s">'SARS'</span><span class="p">,</span> <span class="n">indicate_novel</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">display</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/SARS'</span><span class="p">)</span>
<span class="n">sg</span><span class="p">.</span><span class="n">plot_transcript_path</span><span class="p">(</span><span class="s">'ENCODEHT000445064'</span><span class="p">,</span> <span class="n">indicate_novel</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">display</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/SARS'</span><span class="p">)</span>
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_21_0.png" alt="png" /></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Saving summary graph for ENSG00000031698.12 as figures/SARS_novel_ENSG00000031698.12_summary.png
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_21_2.png" alt="png" /></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Saving transcript path graph for ENCODEHT000445064 as figures/SARS_novel_ENCODEHT000445064_path.png
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># plot gene report for the same gene with IR (SARS)
</span><span class="n">gname</span> <span class="o">=</span> <span class="s">'SARS'</span>
<span class="n">sg</span><span class="p">.</span><span class="n">gen_report</span><span class="p">(</span><span class="n">gname</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/{}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">gname</span><span class="p">),</span>
    <span class="n">heatmap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s">'viridis'</span><span class="p">,</span>
    <span class="n">indicate_novel</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
    <span class="n">novelty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">sg</span><span class="p">.</span><span class="n">gen_report</span><span class="p">(</span><span class="n">gname</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/{}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">gname</span><span class="p">),</span>
    <span class="n">heatmap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s">'viridis'</span><span class="p">,</span>
    <span class="n">browser</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
    <span class="n">novelty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Plotting transcripts for ENSG00000031698.12
Saving transcript path graph for ENST00000234677.6 as figures/SARS_novel_ENST00000234677.6_path.png
Saving transcript path graph for ENST00000369923.4 as figures/SARS_novel_ENST00000369923.4_path.png
Saving transcript path graph for ENCODEHT000444845 as figures/SARS_novel_ENCODEHT000444845_path.png
Saving transcript path graph for ENST00000482384.1 as figures/SARS_novel_ENST00000482384.1_path.png
Saving transcript path graph for ENCODEHT000445064 as figures/SARS_novel_ENCODEHT000445064_path.png
Saving transcript path graph for ENST00000468588.1 as figures/SARS_novel_ENST00000468588.1_path.png
Generating report for ENSG00000031698.12

Plotting transcripts for ENSG00000031698.12
Saving transcript path graph for ENST00000234677.6 as figures/SARS_browser_ENST00000234677.6_path.png
Saving transcript path graph for ENST00000369923.4 as figures/SARS_browser_ENST00000369923.4_path.png
Saving transcript path graph for ENCODEHT000444845 as figures/SARS_browser_ENCODEHT000444845_path.png
Saving transcript path graph for ENST00000482384.1 as figures/SARS_browser_ENST00000482384.1_path.png
Saving transcript path graph for ENCODEHT000445064 as figures/SARS_browser_ENCODEHT000445064_path.png
Saving transcript path graph for ENST00000468588.1 as figures/SARS_browser_ENST00000468588.1_path.png
Generating report for ENSG00000031698.12
</code></pre></div></div>

<p>(Forgive the terrible sample labels, this will be fixed in a future Swan release!)</p>

<p><img src="/assets/images/long-read/sars_swan.png" alt="png" /></p>

<p><img src="/assets/images/long-read/sars_browser.png" alt="png" /></p>

<h3 id="tss-and-isoform-switching">TSS and isoform switching</h3>

<p>Detect statistically-significant instances of isoform or TSS switching from the TALON isoforms or from the called TSSs respectively.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># df: talon abundance file (either tss or transcript)
# cond_map: dictionary of {condition: [dataset1, dataset2]}; how you want to group datasets
# how: whether to make a tss or iso level adata; 'tss' or 'iso'
# pass_list: if 'iso', file of valid transcript IDs that pass filtering
</span><span class="k">def</span> <span class="nf">make_adata</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cond_map</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'iso'</span><span class="p">,</span> <span class="n">pass_list</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

    <span class="c1"># filter talon ab file based on pass list
#     df = pd.read_csv(ab_file, sep='\t')
</span>    <span class="k">if</span> <span class="n">pass_list</span><span class="p">:</span>
        <span class="n">pass_list</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pass_list</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'gene_id'</span><span class="p">,</span> <span class="s">'transcript_id'</span><span class="p">])</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="n">transcript_ID</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">pass_list</span><span class="p">.</span><span class="n">transcript_id</span><span class="p">.</span><span class="n">tolist</span><span class="p">())]</span>

    <span class="c1"># obs table
</span>    <span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">cond_map</span><span class="p">,</span> <span class="n">orient</span><span class="o">=</span><span class="s">'index'</span><span class="p">)</span>
    <span class="n">obs</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">id_vars</span> <span class="o">=</span> <span class="p">[</span><span class="s">'index'</span><span class="p">]</span>
    <span class="n">value_vars</span> <span class="o">=</span> <span class="n">obs</span><span class="p">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="n">obs</span> <span class="o">=</span> <span class="n">obs</span><span class="p">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="n">id_vars</span><span class="p">,</span> <span class="n">value_vars</span><span class="o">=</span><span class="n">value_vars</span><span class="p">)</span>
    <span class="n">obs</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'variable'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">obs</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'index'</span><span class="p">:</span><span class="s">'condition'</span><span class="p">,</span> <span class="s">'value'</span><span class="p">:</span><span class="s">'dataset'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># var table
</span>    <span class="k">if</span> <span class="n">how</span><span class="o">==</span><span class="s">'iso'</span><span class="p">:</span>
        <span class="n">var_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s">'annot_transcript_name'</span><span class="p">,</span> <span class="s">'annot_gene_name'</span><span class="p">,</span>\
                    <span class="s">'annot_transcript_id'</span><span class="p">,</span> <span class="s">'annot_gene_id'</span><span class="p">,</span> \
                    <span class="s">'gene_ID'</span><span class="p">,</span> <span class="s">'transcript_ID'</span><span class="p">,</span> <span class="s">'transcript_novelty'</span><span class="p">,</span> \
                    <span class="s">'ISM_subtype'</span><span class="p">]</span>
        <span class="n">var</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">var_cols</span><span class="p">]</span>
        <span class="n">var</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'transcript_ID'</span><span class="p">:</span><span class="s">'transcript_id'</span><span class="p">,</span> \
                    <span class="s">'gene_ID'</span><span class="p">:</span><span class="s">'gene_id'</span><span class="p">,</span>\
                    <span class="s">'annot_gene_name'</span><span class="p">:</span> <span class="s">'gene_name'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">how</span><span class="o">==</span><span class="s">'tss'</span><span class="p">:</span>
        <span class="n">var_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s">'annot_gene_name'</span><span class="p">,</span> <span class="s">'tss_id'</span><span class="p">]</span>
        <span class="n">var</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">var_cols</span><span class="p">]</span>
        <span class="n">var</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'annot_gene_name'</span><span class="p">:</span> <span class="s">'gene_name'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># X table
</span>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="n">index</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">obs</span><span class="p">.</span><span class="n">dataset</span><span class="p">.</span><span class="n">tolist</span><span class="p">())]</span>
    <span class="n">obs_order</span> <span class="o">=</span> <span class="n">obs</span><span class="p">[</span><span class="s">'dataset'</span><span class="p">].</span><span class="n">reset_index</span><span class="p">().</span><span class="n">set_index</span><span class="p">(</span><span class="s">'dataset'</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s">'dataset_num'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">index</span><span class="p">.</span><span class="nb">map</span><span class="p">(</span><span class="n">obs_order</span><span class="p">[</span><span class="s">'index'</span><span class="p">])</span>
    <span class="n">df</span><span class="p">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s">'dataset_num'</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'dataset_num'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">to_numpy</span><span class="p">()</span>

    <span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="p">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">obs</span><span class="o">=</span><span class="n">obs</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">var</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">X</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">adata</span>

<span class="c1"># gene_df: pandas dataframe with expression values in each condition for each TSS in a gene
# conditions: list of str of condition names
# rc: threshold of read count per gene in each condition necessary to test this gene
</span><span class="k">def</span> <span class="nf">test_gene</span><span class="p">(</span><span class="n">gene_df</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">id_col</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>

    <span class="n">gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">col</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">id_col</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s">'counts'</span><span class="p">)</span>
    <span class="n">gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">transpose</span><span class="p">()</span>

    <span class="n">groups</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">columns</span><span class="p">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">gene_df</span><span class="p">[</span><span class="s">'total_counts'</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">groups</span><span class="p">].</span><span class="nb">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">gene_df</span><span class="p">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s">'total_counts'</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_df</span><span class="p">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">11</span><span class="p">:</span>
        <span class="n">gene_df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

        <span class="n">beep</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">10</span><span class="p">:].</span><span class="nb">sum</span><span class="p">()</span>
        <span class="n">beep</span><span class="p">[</span><span class="n">id_col</span><span class="p">]</span> <span class="o">=</span> <span class="s">'all_other'</span>
        <span class="n">beep</span><span class="p">.</span><span class="n">index</span><span class="p">.</span><span class="n">name</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">beep</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">beep</span><span class="p">).</span><span class="n">transpose</span><span class="p">()</span>

        <span class="n">gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">10</span><span class="p">]</span>
        <span class="n">gene_df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">concat</span><span class="p">([</span><span class="n">gene_df</span><span class="p">,</span> <span class="n">beep</span><span class="p">])</span>

    <span class="c1"># limit to just isoforms with &gt; 0 expression in both conditions
</span>    <span class="n">cond1</span> <span class="o">=</span> <span class="n">conditions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">cond2</span> <span class="o">=</span> <span class="n">conditions</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">gene_df</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">gene_df</span><span class="p">[</span><span class="n">cond1</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">gene_df</span><span class="p">[</span><span class="n">cond2</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">)]</span>

    <span class="c1"># does this gene reach the desired read count threshold?
</span>    <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">cond</span><span class="p">].</span><span class="nb">sum</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">rc</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="p">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="p">.</span><span class="n">nan</span>

    <span class="c1"># only do the rest if there's nothing left
</span>    <span class="k">if</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="p">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="p">.</span><span class="n">nan</span>

    <span class="c1"># calculate the percent of each sample each TSS accounts for
</span>    <span class="n">cond_pis</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="n">total_col</span> <span class="o">=</span> <span class="s">'{}_total'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">cond</span><span class="p">)</span>
        <span class="n">pi_col</span> <span class="o">=</span> <span class="s">'{}_pi'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">cond</span><span class="p">)</span>
        <span class="n">total_count</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">cond</span><span class="p">].</span><span class="nb">sum</span><span class="p">()</span>

        <span class="n">cond_pis</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">pi_col</span><span class="p">)</span>

        <span class="n">gene_df</span><span class="p">[</span><span class="n">total_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">total_count</span>
        <span class="n">gene_df</span><span class="p">[</span><span class="n">pi_col</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">gene_df</span><span class="p">[</span><span class="n">cond</span><span class="p">]</span><span class="o">/</span><span class="n">gene_df</span><span class="p">[</span><span class="n">total_col</span><span class="p">])</span><span class="o">*</span><span class="mi">100</span>

    <span class="c1"># compute isoform-level and gene-level delta pis
</span>    <span class="n">gene_df</span><span class="p">[</span><span class="s">'dpi'</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">cond_pis</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">-</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">cond_pis</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
    <span class="n">gene_df</span><span class="p">[</span><span class="s">'abs_dpi'</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">dpi</span><span class="p">.</span><span class="nb">abs</span><span class="p">()</span>
    <span class="n">gene_dpi</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">2</span><span class="p">].</span><span class="n">abs_dpi</span><span class="p">.</span><span class="nb">sum</span><span class="p">()</span>

    <span class="c1"># chi squared test
</span>    <span class="n">chi_table</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">conditions</span><span class="p">].</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">chi2</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">dof</span><span class="p">,</span> <span class="n">exp</span> <span class="o">=</span> <span class="n">st</span><span class="p">.</span><span class="n">chi2_contingency</span><span class="p">(</span><span class="n">chi_table</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">p</span><span class="p">,</span> <span class="n">gene_dpi</span>

<span class="k">def</span> <span class="nf">filter_die_results</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">dpi</span><span class="p">):</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span><span class="p">.</span><span class="n">adj_p_val</span><span class="o">&lt;=</span><span class="n">p</span><span class="p">)</span><span class="o">&amp;</span><span class="p">(</span><span class="n">df</span><span class="p">.</span><span class="n">dpi</span><span class="o">&gt;=</span><span class="n">dpi</span><span class="p">)]</span>
    <span class="k">return</span> <span class="n">df</span>

<span class="c1"># adata: adata with TSS or iso expression
# conditions: len 2 list of strings of conditions to compare
# col: string, which column the condition labels are in
# how: 'tss' or 'iso'
</span><span class="k">def</span> <span class="nf">get_die</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'tss'</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="mi">15</span><span class="p">):</span>

    <span class="k">if</span> <span class="n">how</span> <span class="o">==</span> <span class="s">'tss'</span><span class="p">:</span>
        <span class="n">id_col</span> <span class="o">=</span> <span class="s">'tss_id'</span>
    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s">'iso'</span><span class="p">:</span>
        <span class="n">id_col</span> <span class="o">=</span> <span class="s">'transcript_id'</span>

    <span class="c1"># make df that we can groupby
</span>    <span class="n">col</span> <span class="o">=</span> <span class="s">'condition'</span>
    <span class="n">colnames</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">var</span><span class="p">[</span><span class="n">id_col</span><span class="p">].</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">rownames</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">obs</span><span class="p">.</span><span class="n">dataset</span><span class="p">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">X</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">raw</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rownames</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">colnames</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'index'</span><span class="p">:</span><span class="s">'dataset'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">samp</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">obs</span><span class="p">[[</span><span class="s">'dataset'</span><span class="p">,</span> <span class="n">col</span><span class="p">]]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">samp</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s">'dataset'</span><span class="p">)</span>

    <span class="c1"># limit to only the samples that we want in this condition
</span>    <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">].</span><span class="n">astype</span><span class="p">(</span><span class="s">'str'</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">].</span><span class="n">isin</span><span class="p">(</span><span class="n">conditions</span><span class="p">)]</span>

    <span class="c1"># groupby sample type and sum over gen
</span>    <span class="n">df</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'dataset'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">col</span><span class="p">).</span><span class="nb">sum</span><span class="p">().</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1"># melty boi
</span>    <span class="n">tss_cols</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">columns</span><span class="p">.</span><span class="n">tolist</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="n">col</span><span class="p">,</span> <span class="n">value_vars</span><span class="o">=</span><span class="n">tss_cols</span><span class="p">)</span>

    <span class="c1"># rename some cols
</span>    <span class="n">df</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'variable'</span><span class="p">:</span><span class="n">id_col</span><span class="p">,</span><span class="s">'value'</span><span class="p">:</span><span class="s">'counts'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># merge with gene names
</span>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata</span><span class="p">.</span><span class="n">var</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="n">id_col</span><span class="p">)</span>

    <span class="c1"># get total number of tss or iso / gene
</span>    <span class="n">bop</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s">'gene_name'</span><span class="p">,</span> <span class="n">id_col</span><span class="p">]].</span><span class="n">groupby</span><span class="p">(</span><span class="s">'gene_name'</span><span class="p">).</span><span class="n">count</span><span class="p">().</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1"># construct tables for each gene and test!
</span>    <span class="n">gene_names</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">gene_name</span><span class="p">.</span><span class="n">unique</span><span class="p">().</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">gene_de_df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">gene_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s">'p_val'</span><span class="p">,</span> <span class="s">'dpi'</span><span class="p">],</span> <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="n">np</span><span class="p">.</span><span class="n">nan</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)]</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_names</span><span class="p">))])</span>
    <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_names</span><span class="p">:</span>
        <span class="n">gene_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">.</span><span class="n">gene_name</span><span class="o">==</span><span class="n">gene</span><span class="p">]</span>
        <span class="n">p</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="n">test_gene</span><span class="p">(</span><span class="n">gene_df</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">col</span><span class="p">,</span> <span class="n">id_col</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="n">rc</span><span class="p">)</span>
        <span class="n">gene_de_df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s">'p_val'</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span>
        <span class="n">gene_de_df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="s">'dpi'</span><span class="p">]</span> <span class="o">=</span> <span class="n">dpi</span>

    <span class="c1"># correct p values
</span>    <span class="n">gene_de_df</span><span class="p">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">p_vals</span> <span class="o">=</span> <span class="n">gene_de_df</span><span class="p">.</span><span class="n">p_val</span><span class="p">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">_</span><span class="p">,</span> <span class="n">adj_p_vals</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">multipletests</span><span class="p">(</span><span class="n">p_vals</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s">'fdr_bh'</span><span class="p">)</span>
    <span class="n">gene_de_df</span><span class="p">[</span><span class="s">'adj_p_val'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adj_p_vals</span>

    <span class="n">gene_de_df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">gene_de_df</span>

<span class="k">def</span> <span class="nf">get_sample_colors</span><span class="p">():</span>
    <span class="n">c_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s">'astro_1'</span><span class="p">:</span> <span class="s">'#f6ef7c'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">:</span> <span class="s">'#eabc68'</span><span class="p">,</span>\
          <span class="s">'excite_neuron_1'</span><span class="p">:</span> <span class="s">'#e4d3cd'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">:</span> <span class="s">'#d3a8b2'</span><span class="p">,</span>\
          <span class="s">'pgp1_1'</span><span class="p">:</span> <span class="s">'#bef4ff'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">:</span> <span class="s">'#73a8b2'</span><span class="p">}</span>
    <span class="n">order</span> <span class="o">=</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">,</span> <span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">,</span> <span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span>

<span class="k">def</span> <span class="nf">make_cond_map</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span> <span class="n">group_names</span><span class="p">):</span>
    <span class="n">cond_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">group_name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span> <span class="n">group_names</span><span class="p">):</span>
        <span class="n">cond_map</span><span class="p">[</span><span class="n">group</span><span class="p">]</span> <span class="o">=</span> <span class="n">group_name</span>
    <span class="k">return</span> <span class="n">cond_map</span>

<span class="c1"># calculate the normalized average or sum of TSS expression
# per cell from the TSS anndata object
</span><span class="k">def</span> <span class="nf">calc_exp</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="n">group_names</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'tss'</span><span class="p">,</span> <span class="n">cpm</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">adata</span><span class="p">.</span><span class="n">var</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="k">if</span> <span class="n">how</span> <span class="o">==</span> <span class="s">'tss'</span><span class="p">:</span>
        <span class="n">id_col</span> <span class="o">=</span> <span class="s">'tss_id'</span>
    <span class="k">elif</span> <span class="n">how</span> <span class="o">==</span> <span class="s">'iso'</span><span class="p">:</span>
        <span class="n">id_col</span> <span class="o">=</span> <span class="s">'transcript_id'</span>

    <span class="c1"># conditions map
</span>    <span class="n">cond_map</span> <span class="o">=</span> <span class="n">make_cond_map</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span> <span class="n">group_names</span><span class="p">)</span>
<span class="c1">#     print(cond_map)
</span>    <span class="n">col</span> <span class="o">=</span> <span class="s">'condition'</span>
    <span class="n">adata</span><span class="p">.</span><span class="n">obs</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'dataset'</span><span class="p">:</span><span class="s">'condition'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="c1">#     adata.obs[col] = adata.obs.dataset.map(cond_map)
</span>
    <span class="c1"># make df that we can groupby
</span>    <span class="n">colnames</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">var</span><span class="p">[</span><span class="n">id_col</span><span class="p">].</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">rownames</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">obs</span><span class="p">[</span><span class="n">col</span><span class="p">].</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">X</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">var</span><span class="p">.</span><span class="n">gene_name</span><span class="p">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">raw</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">rownames</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">colnames</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">df</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'index'</span><span class="p">:</span><span class="n">col</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">samp</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">obs</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">samp</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="n">col</span><span class="p">)</span>

    <span class="c1"># limit to only the cells that we want in this condition
</span>    <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">].</span><span class="n">astype</span><span class="p">(</span><span class="s">'str'</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">].</span><span class="n">isin</span><span class="p">(</span><span class="n">group_names</span><span class="p">)]</span>

    <span class="c1"># groupby sample type and sum over gen
#     df.drop(col, axis=1, inplace=True)
</span>    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">col</span><span class="p">).</span><span class="nb">sum</span><span class="p">().</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">cpm</span><span class="p">:</span>
        <span class="c1"># since these values haven't been normalized yet, do that
</span>        <span class="c1"># CPM : (counts/total_counts)* 1**6
</span>        <span class="c1"># Note : ATAC values were pre-normalized
</span>        <span class="n">df</span><span class="p">.</span><span class="n">set_index</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">transpose</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">group_names</span><span class="p">:</span>
            <span class="n">total_counts</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">].</span><span class="nb">sum</span><span class="p">()</span>
            <span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">/</span><span class="n">total_counts</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="mi">1</span><span class="o">^</span><span class="mi">6</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">transpose</span><span class="p">()</span>
        <span class="n">df</span><span class="p">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># melty boi
</span>    <span class="n">tss_cols</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">columns</span><span class="p">.</span><span class="n">tolist</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="n">col</span><span class="p">,</span> <span class="n">value_vars</span><span class="o">=</span><span class="n">tss_cols</span><span class="p">)</span>

    <span class="c1"># rename some cols
</span>    <span class="n">df</span><span class="p">.</span><span class="n">rename</span><span class="p">({</span><span class="s">'variable'</span><span class="p">:</span><span class="n">id_col</span><span class="p">,</span><span class="s">'value'</span><span class="p">:</span><span class="s">'counts'</span><span class="p">},</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># add gene name
</span>    <span class="k">if</span> <span class="n">how</span> <span class="o">==</span> <span class="s">'tss'</span><span class="p">:</span>
        <span class="n">temp</span> <span class="o">=</span> <span class="n">adata</span><span class="p">.</span><span class="n">var</span><span class="p">[[</span><span class="n">id_col</span><span class="p">,</span> <span class="s">'gene_name'</span><span class="p">]]</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">merge</span><span class="p">(</span><span class="n">temp</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="n">id_col</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span>

<span class="k">def</span> <span class="nf">plot_tss_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="n">group_names</span><span class="p">,</span> <span class="n">gname</span><span class="p">,</span> <span class="n">opref</span><span class="p">):</span>

    <span class="c1"># calculate TSS expression per condition
</span>    <span class="n">adata</span><span class="p">.</span><span class="n">obs</span><span class="p">.</span><span class="n">drop</span><span class="p">(</span><span class="s">'condition'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">calc_exp</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="n">group_names</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'tss'</span><span class="p">)</span>

    <span class="c1"># subset by gene and calculate DPI per gene
</span>    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">tss_df</span><span class="p">.</span><span class="n">loc</span><span class="p">[</span><span class="n">tss_df</span><span class="p">.</span><span class="n">gene_name</span> <span class="o">==</span> <span class="n">gname</span><span class="p">]</span>
    <span class="n">n_tss</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">tss_df</span><span class="p">.</span><span class="n">tss_id</span><span class="p">.</span><span class="n">unique</span><span class="p">())</span>
    <span class="n">tss_df</span><span class="p">.</span><span class="n">drop</span><span class="p">([</span><span class="s">'gene_name'</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">tss_df</span><span class="p">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="s">'tss_id'</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s">'condition'</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="s">'counts'</span><span class="p">)</span>
    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">tss_df</span><span class="p">.</span><span class="n">div</span><span class="p">(</span><span class="n">tss_df</span><span class="p">.</span><span class="nb">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># get categorical colormap
</span>    <span class="n">tss_df</span> <span class="o">=</span> <span class="n">tss_df</span><span class="p">[</span><span class="n">groups</span><span class="p">]</span>
    <span class="n">mini_obs</span><span class="p">,</span> <span class="n">cat_cmap</span> <span class="o">=</span> <span class="n">get_cat_cmap</span><span class="p">(</span><span class="n">groups</span><span class="p">)</span>

    <span class="c1"># plot the figure
</span>    <span class="n">sns</span><span class="p">.</span><span class="nb">set</span><span class="p">(</span><span class="n">rc</span><span class="o">=</span><span class="p">{</span><span class="s">'figure.figsize'</span><span class="p">:(</span><span class="mi">12</span><span class="p">,</span><span class="mi">7</span><span class="p">)})</span>
    <span class="n">sns</span><span class="p">.</span><span class="n">set_context</span><span class="p">(</span><span class="s">'paper'</span><span class="p">,</span> <span class="n">font_scale</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="p">.</span><span class="n">figure</span><span class="p">()</span>


    <span class="c1"># complicated subplot stuff
</span>    <span class="n">tss_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="p">.</span><span class="n">subplot2grid</span><span class="p">((</span><span class="n">n_tss</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="n">rowspan</span><span class="o">=</span><span class="n">n_tss</span><span class="p">)</span>

    <span class="c1"># fig, axes = plt.subplots(nrows=4)
</span>    <span class="n">fig</span><span class="p">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">hspace</span><span class="o">=</span><span class="mf">0.00</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">wspace</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>

    <span class="c1"># plot tss only plot
</span>    <span class="n">sns</span><span class="p">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">tss_df</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s">'magma'</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">tss_ax</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">tss_ax</span><span class="p">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">tss_ax</span><span class="p">.</span><span class="n">get_yticklabels</span><span class="p">(),</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">tss_ax</span><span class="p">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>
    <span class="n">tss_ax</span><span class="p">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>

    <span class="c1"># plot sample labels
</span>    <span class="n">tss_colorbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="p">.</span><span class="n">add_subplot</span><span class="p">((</span><span class="n">n_tss</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,((</span><span class="n">n_tss</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="mi">2</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">sns</span><span class="p">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">mini_obs</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cat_cmap</span><span class="p">,</span>
                <span class="n">ax</span><span class="o">=</span><span class="n">tss_colorbar_ax</span><span class="p">,</span> <span class="n">cbar</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
    <span class="n">tss_colorbar_ax</span><span class="p">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>
    <span class="n">tss_colorbar_ax</span><span class="p">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>
    <span class="n">tss_colorbar_ax</span><span class="p">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">labelleft</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">tss_colorbar_ax</span><span class="p">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">right</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">labelright</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">tss_colorbar_ax</span><span class="p">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>

    <span class="c1"># plot colorbars
</span>    <span class="n">tss_colorbar_ax</span> <span class="o">=</span> <span class="n">fig</span><span class="p">.</span><span class="n">add_subplot</span><span class="p">((</span><span class="n">n_tss</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="mi">5</span><span class="p">,</span><span class="mi">1</span><span class="p">,((</span><span class="n">n_tss</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="mi">5</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="p">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="s">'magma'</span><span class="p">)</span>
    <span class="n">norm</span> <span class="o">=</span> <span class="n">mpl</span><span class="p">.</span><span class="n">colors</span><span class="p">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">cb</span> <span class="o">=</span> <span class="n">mpl</span><span class="p">.</span><span class="n">colorbar</span><span class="p">.</span><span class="n">ColorbarBase</span><span class="p">(</span><span class="n">tss_colorbar_ax</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>
                                  <span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s">'horizontal'</span><span class="p">)</span>
    <span class="n">cb</span><span class="p">.</span><span class="n">set_label</span><span class="p">(</span><span class="s">'Proportion TSS usage'</span><span class="p">)</span>

    <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_{}_heatmap.pdf'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">opref</span><span class="p">,</span> <span class="n">gname</span><span class="p">)</span>
    <span class="n">plt</span><span class="p">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s">'tight'</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">get_cat_cmap</span><span class="p">(</span><span class="n">sample_order</span><span class="p">):</span>

    <span class="c1"># get condition and cluster colors
</span>    <span class="n">samp_cdict</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">get_sample_colors</span><span class="p">()</span>
    <span class="n">samples</span> <span class="o">=</span> <span class="n">sample_order</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="p">.</span><span class="n">transpose</span><span class="p">([</span><span class="n">samples</span><span class="p">])</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
                      <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s">'sample'</span><span class="p">])</span>

    <span class="c1"># assign arbitrary numbers to each category (cluster, condition)
</span>    <span class="n">cats</span> <span class="o">=</span> <span class="n">samples</span>
    <span class="n">cat_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">cat</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cat</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cats</span><span class="p">)])</span>
    <span class="n">df</span><span class="p">[</span><span class="s">'sample'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s">'sample'</span><span class="p">].</span><span class="nb">map</span><span class="p">(</span><span class="n">cat_dict</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="n">transpose</span><span class="p">()</span>

    <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">samp_cdict</span><span class="p">[</span><span class="n">cat</span><span class="p">]</span> <span class="k">for</span> <span class="n">cat</span> <span class="ow">in</span> <span class="n">cats</span><span class="p">]</span>
    <span class="n">cat_cmap</span> <span class="o">=</span> <span class="n">ListedColormap</span><span class="p">(</span><span class="n">colors</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span><span class="p">,</span> <span class="n">cat_cmap</span>
</code></pre></div></div>

<h3 id="tss-switching">TSS switching</h3>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ab_file</span> <span class="o">=</span> <span class="s">'pgp1_tss_talon_abundance.tsv'</span>
<span class="n">cond_map</span> <span class="o">=</span> <span class="p">{</span><span class="s">'Astrocytes'</span><span class="p">:</span> <span class="p">[</span><span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">],</span> \
            <span class="s">'Excitatory neurons'</span><span class="p">:</span> <span class="p">[</span><span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">],</span> \
            <span class="s">'PGP1'</span><span class="p">:</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">]}</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ab_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span><span class="n">a</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">make_adata</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cond_map</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'tss'</span><span class="p">)</span>

<span class="n">fname</span> <span class="o">=</span> <span class="s">'tss.h5ad'</span>
<span class="n">adata</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>

<span class="c1"># do one test for each pair of conditions
</span><span class="n">tested</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">conditions</span> <span class="o">=</span> <span class="p">[</span><span class="s">'Astrocytes'</span><span class="p">,</span> <span class="s">'Excitatory neurons'</span><span class="p">,</span> <span class="s">'PGP1'</span><span class="p">]</span>
<span class="n">how</span> <span class="o">=</span> <span class="s">'tss'</span>
<span class="n">p</span> <span class="o">=</span> <span class="mf">0.05</span> <span class="c1"># adjusted p-value threshold
</span><span class="n">dpi</span> <span class="o">=</span> <span class="mi">10</span> <span class="c1"># change in percent usage (dpi) threshold
</span><span class="k">for</span> <span class="n">c1</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">c2</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tested</span> <span class="ow">or</span> <span class="n">c1</span> <span class="o">==</span> <span class="n">c2</span> <span class="ow">or</span> <span class="p">(</span><span class="n">c2</span><span class="p">,</span> <span class="n">c1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tested</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tested</span><span class="p">.</span><span class="n">append</span><span class="p">((</span><span class="n">c1</span><span class="p">,</span><span class="n">c2</span><span class="p">))</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">get_die</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="n">how</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

        <span class="c1"># all results
</span>        <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_{}_{}_die.tsv'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">,</span> <span class="n">how</span><span class="p">)</span>
        <span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="c1"># significant results
</span>        <span class="n">df</span> <span class="o">=</span> <span class="n">filter_die_results</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">dpi</span><span class="p">)</span>
        <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_{}_{}_sig_die.tsv'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">,</span> <span class="n">how</span><span class="p">)</span>
        <span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># visualize some TSS switches that were called
</span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="p">.</span><span class="n">read</span><span class="p">(</span><span class="s">'tss.h5ad'</span><span class="p">)</span>
<span class="n">groups</span> <span class="o">=</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">,</span> <span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">,</span> <span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">]</span>

<span class="n">tss_df</span> <span class="o">=</span> <span class="n">plot_tss_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="s">'TCF4'</span><span class="p">,</span> <span class="s">'figures/tss'</span><span class="p">)</span>

<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="p">.</span><span class="n">read</span><span class="p">(</span><span class="s">'tss.h5ad'</span><span class="p">)</span>
<span class="n">tss_df</span> <span class="o">=</span> <span class="n">plot_tss_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="n">groups</span><span class="p">,</span> <span class="s">'CALD1'</span><span class="p">,</span> <span class="s">'figures/tss'</span><span class="p">)</span>
</code></pre></div></div>

<p><img src="/assets/images/long-read/output_27_0.png" alt="png" /></p>

<p><img src="/assets/images/long-read/output_27_1.png" alt="png" /></p>

<h3 id="isoform-switching">Isoform switching</h3>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ab_file</span> <span class="o">=</span> <span class="s">'pgp1_tss_talon_abundance.tsv'</span>
<span class="n">pass_list</span> <span class="o">=</span> <span class="s">'pgp1_pass_list.csv'</span>
<span class="n">cond_map</span> <span class="o">=</span> <span class="p">{</span><span class="s">'Astrocytes'</span><span class="p">:</span> <span class="p">[</span><span class="s">'astro_1'</span><span class="p">,</span> <span class="s">'astro_2'</span><span class="p">],</span> \
            <span class="s">'Excitatory neurons'</span><span class="p">:</span> <span class="p">[</span><span class="s">'excite_neuron_1'</span><span class="p">,</span> <span class="s">'excite_neuron_2'</span><span class="p">],</span> \
            <span class="s">'PGP1'</span><span class="p">:</span> <span class="p">[</span><span class="s">'pgp1_1'</span><span class="p">,</span> <span class="s">'pgp1_2'</span><span class="p">]}</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ab_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">make_adata</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">cond_map</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s">'tss'</span><span class="p">)</span>

<span class="n">fname</span> <span class="o">=</span> <span class="s">'tss.h5ad'</span>
<span class="n">adata</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>

<span class="c1"># do one test for each pair of conditions
</span><span class="n">tested</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">conditions</span> <span class="o">=</span> <span class="p">[</span><span class="s">'Astrocytes'</span><span class="p">,</span> <span class="s">'Excitatory neurons'</span><span class="p">,</span> <span class="s">'PGP1'</span><span class="p">]</span>
<span class="n">how</span> <span class="o">=</span> <span class="s">'tss'</span>
<span class="n">p</span> <span class="o">=</span> <span class="mf">0.05</span> <span class="c1"># adjusted p-value threshold
</span><span class="n">dpi</span> <span class="o">=</span> <span class="mi">10</span> <span class="c1"># change in percent usage (dpi) threshold
</span><span class="k">for</span> <span class="n">c1</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">c2</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tested</span> <span class="ow">or</span> <span class="n">c1</span> <span class="o">==</span> <span class="n">c2</span> <span class="ow">or</span> <span class="p">(</span><span class="n">c2</span><span class="p">,</span> <span class="n">c1</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tested</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">tested</span><span class="p">.</span><span class="n">append</span><span class="p">((</span><span class="n">c1</span><span class="p">,</span><span class="n">c2</span><span class="p">))</span>
            <span class="k">print</span><span class="p">(</span><span class="s">'Testing {} vs. {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">))</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">get_die</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="n">how</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

        <span class="c1"># all results
</span>        <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_{}_{}_die.tsv'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">,</span> <span class="n">how</span><span class="p">)</span>
        <span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

        <span class="c1"># significant results
</span>        <span class="n">df</span> <span class="o">=</span> <span class="n">filter_die_results</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">dpi</span><span class="p">)</span>
        <span class="n">fname</span> <span class="o">=</span> <span class="s">'{}_{}_{}_sig_die.tsv'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">c1</span><span class="p">,</span> <span class="n">c2</span><span class="p">,</span> <span class="n">how</span><span class="p">)</span>
        <span class="n">df</span><span class="p">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># use Swan to visualize some isoform switches
</span><span class="n">sg</span> <span class="o">=</span> <span class="n">swan</span><span class="p">.</span><span class="n">SwanGraph</span><span class="p">(</span><span class="s">'swan.p'</span><span class="p">)</span>
<span class="n">gnames</span> <span class="o">=</span> <span class="p">[</span><span class="s">'TPD52L2'</span><span class="p">,</span> <span class="s">'AP2M1'</span><span class="p">,</span> <span class="s">'SRSF3'</span><span class="p">,</span> <span class="s">'SRSF7'</span><span class="p">]</span>
<span class="k">for</span> <span class="n">gname</span> <span class="ow">in</span> <span class="n">gnames</span><span class="p">:</span>
    <span class="c1"># dpi swangraph
</span>    <span class="n">sg</span><span class="p">.</span><span class="n">gen_report</span><span class="p">(</span><span class="n">gname</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/{}_dpi'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">gname</span><span class="p">),</span>
        <span class="n">heatmap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">dpi</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s">'magma'</span><span class="p">,</span>
        <span class="n">indicate_novel</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">novelty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># tpm swangraph
</span>    <span class="n">sg</span><span class="p">.</span><span class="n">gen_report</span><span class="p">(</span><span class="n">gname</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/{}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">gname</span><span class="p">),</span>
        <span class="n">heatmap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s">'viridis'</span><span class="p">,</span>
        <span class="n">indicate_novel</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">novelty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># dpi browser
</span>    <span class="n">sg</span><span class="p">.</span><span class="n">gen_report</span><span class="p">(</span><span class="n">gname</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/{}_dpi'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">gname</span><span class="p">),</span>
        <span class="n">heatmap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">dpi</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s">'magma'</span><span class="p">,</span>
        <span class="n">browser</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">novelty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

    <span class="c1"># tpm browser
</span>    <span class="n">sg</span><span class="p">.</span><span class="n">gen_report</span><span class="p">(</span><span class="n">gname</span><span class="p">,</span> <span class="n">prefix</span><span class="o">=</span><span class="s">'figures/{}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">gname</span><span class="p">),</span>
        <span class="n">heatmap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">cmap</span><span class="o">=</span><span class="s">'viridis'</span><span class="p">,</span>
        <span class="n">browser</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
        <span class="n">novelty</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Graph from swan.p loaded

Plotting transcripts for ENSG00000101150.17
Saving transcript path graph for ENST00000346249.8 as figures/TPD52L2_dpi_novel_ENST00000346249.8_path.png
Saving transcript path graph for ENST00000348257.9 as figures/TPD52L2_dpi_novel_ENST00000348257.9_path.png
Saving transcript path graph for ENST00000358548.4 as figures/TPD52L2_dpi_novel_ENST00000358548.4_path.png
Saving transcript path graph for ENST00000352482.8 as figures/TPD52L2_dpi_novel_ENST00000352482.8_path.png
Saving transcript path graph for ENST00000217121.9 as figures/TPD52L2_dpi_novel_ENST00000217121.9_path.png
Saving transcript path graph for ENST00000351424.8 as figures/TPD52L2_dpi_novel_ENST00000351424.8_path.png
Saving transcript path graph for ENST00000474176.2 as figures/TPD52L2_dpi_novel_ENST00000474176.2_path.png
Generating report for ENSG00000101150.17

Plotting transcripts for ENSG00000101150.17
Saving transcript path graph for ENST00000346249.8 as figures/TPD52L2_novel_ENST00000346249.8_path.png
Saving transcript path graph for ENST00000348257.9 as figures/TPD52L2_novel_ENST00000348257.9_path.png
Saving transcript path graph for ENST00000358548.4 as figures/TPD52L2_novel_ENST00000358548.4_path.png
Saving transcript path graph for ENST00000352482.8 as figures/TPD52L2_novel_ENST00000352482.8_path.png
Saving transcript path graph for ENST00000217121.9 as figures/TPD52L2_novel_ENST00000217121.9_path.png
Saving transcript path graph for ENST00000351424.8 as figures/TPD52L2_novel_ENST00000351424.8_path.png
Saving transcript path graph for ENST00000474176.2 as figures/TPD52L2_novel_ENST00000474176.2_path.png
Generating report for ENSG00000101150.17

Plotting transcripts for ENSG00000101150.17
Saving transcript path graph for ENST00000346249.8 as figures/TPD52L2_dpi_browser_ENST00000346249.8_path.png
Saving transcript path graph for ENST00000348257.9 as figures/TPD52L2_dpi_browser_ENST00000348257.9_path.png
Saving transcript path graph for ENST00000358548.4 as figures/TPD52L2_dpi_browser_ENST00000358548.4_path.png
Saving transcript path graph for ENST00000352482.8 as figures/TPD52L2_dpi_browser_ENST00000352482.8_path.png
Saving transcript path graph for ENST00000217121.9 as figures/TPD52L2_dpi_browser_ENST00000217121.9_path.png
Saving transcript path graph for ENST00000351424.8 as figures/TPD52L2_dpi_browser_ENST00000351424.8_path.png
Saving transcript path graph for ENST00000474176.2 as figures/TPD52L2_dpi_browser_ENST00000474176.2_path.png
Generating report for ENSG00000101150.17

Plotting transcripts for ENSG00000101150.17
Saving transcript path graph for ENST00000346249.8 as figures/TPD52L2_browser_ENST00000346249.8_path.png
Saving transcript path graph for ENST00000348257.9 as figures/TPD52L2_browser_ENST00000348257.9_path.png
Saving transcript path graph for ENST00000358548.4 as figures/TPD52L2_browser_ENST00000358548.4_path.png
Saving transcript path graph for ENST00000352482.8 as figures/TPD52L2_browser_ENST00000352482.8_path.png
Saving transcript path graph for ENST00000217121.9 as figures/TPD52L2_browser_ENST00000217121.9_path.png
Saving transcript path graph for ENST00000351424.8 as figures/TPD52L2_browser_ENST00000351424.8_path.png
Saving transcript path graph for ENST00000474176.2 as figures/TPD52L2_browser_ENST00000474176.2_path.png
Generating report for ENSG00000101150.17

Plotting transcripts for ENSG00000161203.13
Saving transcript path graph for ENST00000382456.7 as figures/AP2M1_dpi_novel_ENST00000382456.7_path.png
Saving transcript path graph for ENCODEHT000422582 as figures/AP2M1_dpi_novel_ENCODEHT000422582_path.png
Saving transcript path graph for ENST00000439647.5 as figures/AP2M1_dpi_novel_ENST00000439647.5_path.png
Saving transcript path graph for ENST00000461733.5 as figures/AP2M1_dpi_novel_ENST00000461733.5_path.png
Saving transcript path graph for ENST00000466598.5 as figures/AP2M1_dpi_novel_ENST00000466598.5_path.png
Generating report for ENSG00000161203.13

Plotting transcripts for ENSG00000161203.13
Saving transcript path graph for ENST00000382456.7 as figures/AP2M1_novel_ENST00000382456.7_path.png
Saving transcript path graph for ENCODEHT000422582 as figures/AP2M1_novel_ENCODEHT000422582_path.png
Saving transcript path graph for ENST00000439647.5 as figures/AP2M1_novel_ENST00000439647.5_path.png
Saving transcript path graph for ENST00000461733.5 as figures/AP2M1_novel_ENST00000461733.5_path.png
Saving transcript path graph for ENST00000466598.5 as figures/AP2M1_novel_ENST00000466598.5_path.png
Generating report for ENSG00000161203.13

Plotting transcripts for ENSG00000161203.13
Saving transcript path graph for ENST00000382456.7 as figures/AP2M1_dpi_browser_ENST00000382456.7_path.png
Saving transcript path graph for ENCODEHT000422582 as figures/AP2M1_dpi_browser_ENCODEHT000422582_path.png
Saving transcript path graph for ENST00000439647.5 as figures/AP2M1_dpi_browser_ENST00000439647.5_path.png
Saving transcript path graph for ENST00000461733.5 as figures/AP2M1_dpi_browser_ENST00000461733.5_path.png
Saving transcript path graph for ENST00000466598.5 as figures/AP2M1_dpi_browser_ENST00000466598.5_path.png
Generating report for ENSG00000161203.13

Plotting transcripts for ENSG00000161203.13
Saving transcript path graph for ENST00000382456.7 as figures/AP2M1_browser_ENST00000382456.7_path.png
Saving transcript path graph for ENCODEHT000422582 as figures/AP2M1_browser_ENCODEHT000422582_path.png
Saving transcript path graph for ENST00000439647.5 as figures/AP2M1_browser_ENST00000439647.5_path.png
Saving transcript path graph for ENST00000461733.5 as figures/AP2M1_browser_ENST00000461733.5_path.png
Saving transcript path graph for ENST00000466598.5 as figures/AP2M1_browser_ENST00000466598.5_path.png
Generating report for ENSG00000161203.13

Plotting transcripts for ENSG00000112081.16
Saving transcript path graph for ENST00000373715.10 as figures/SRSF3_dpi_novel_ENST00000373715.10_path.png
Saving transcript path graph for ENST00000477442.6 as figures/SRSF3_dpi_novel_ENST00000477442.6_path.png
Saving transcript path graph for ENST00000620389.1 as figures/SRSF3_dpi_novel_ENST00000620389.1_path.png
Saving transcript path graph for ENST00000614136.1 as figures/SRSF3_dpi_novel_ENST00000614136.1_path.png
Saving transcript path graph for ENST00000339436.11 as figures/SRSF3_dpi_novel_ENST00000339436.11_path.png
Generating report for ENSG00000112081.16

Plotting transcripts for ENSG00000112081.16
Saving transcript path graph for ENST00000373715.10 as figures/SRSF3_novel_ENST00000373715.10_path.png
Saving transcript path graph for ENST00000477442.6 as figures/SRSF3_novel_ENST00000477442.6_path.png
Saving transcript path graph for ENST00000620389.1 as figures/SRSF3_novel_ENST00000620389.1_path.png
Saving transcript path graph for ENST00000614136.1 as figures/SRSF3_novel_ENST00000614136.1_path.png
Saving transcript path graph for ENST00000339436.11 as figures/SRSF3_novel_ENST00000339436.11_path.png
Generating report for ENSG00000112081.16

Plotting transcripts for ENSG00000112081.16
Saving transcript path graph for ENST00000373715.10 as figures/SRSF3_dpi_browser_ENST00000373715.10_path.png
Saving transcript path graph for ENST00000477442.6 as figures/SRSF3_dpi_browser_ENST00000477442.6_path.png
Saving transcript path graph for ENST00000620389.1 as figures/SRSF3_dpi_browser_ENST00000620389.1_path.png
Saving transcript path graph for ENST00000614136.1 as figures/SRSF3_dpi_browser_ENST00000614136.1_path.png
Saving transcript path graph for ENST00000339436.11 as figures/SRSF3_dpi_browser_ENST00000339436.11_path.png
Generating report for ENSG00000112081.16

Plotting transcripts for ENSG00000112081.16
Saving transcript path graph for ENST00000373715.10 as figures/SRSF3_browser_ENST00000373715.10_path.png
Saving transcript path graph for ENST00000477442.6 as figures/SRSF3_browser_ENST00000477442.6_path.png
Saving transcript path graph for ENST00000620389.1 as figures/SRSF3_browser_ENST00000620389.1_path.png
Saving transcript path graph for ENST00000614136.1 as figures/SRSF3_browser_ENST00000614136.1_path.png
Saving transcript path graph for ENST00000339436.11 as figures/SRSF3_browser_ENST00000339436.11_path.png
Generating report for ENSG00000112081.16

Plotting transcripts for ENSG00000115875.18
Saving transcript path graph for ENST00000313117.10 as figures/SRSF7_dpi_novel_ENST00000313117.10_path.png
Saving transcript path graph for ENST00000446327.6 as figures/SRSF7_dpi_novel_ENST00000446327.6_path.png
Saving transcript path graph for ENST00000443213.5 as figures/SRSF7_dpi_novel_ENST00000443213.5_path.png
Saving transcript path graph for ENST00000409276.5 as figures/SRSF7_dpi_novel_ENST00000409276.5_path.png
Saving transcript path graph for ENST00000425778.5 as figures/SRSF7_dpi_novel_ENST00000425778.5_path.png
Saving transcript path graph for ENST00000431066.5 as figures/SRSF7_dpi_novel_ENST00000431066.5_path.png
Saving transcript path graph for ENST00000477635.5 as figures/SRSF7_dpi_novel_ENST00000477635.5_path.png
Generating report for ENSG00000115875.18

Plotting transcripts for ENSG00000115875.18
Saving transcript path graph for ENST00000313117.10 as figures/SRSF7_novel_ENST00000313117.10_path.png
Saving transcript path graph for ENST00000446327.6 as figures/SRSF7_novel_ENST00000446327.6_path.png
Saving transcript path graph for ENST00000443213.5 as figures/SRSF7_novel_ENST00000443213.5_path.png
Saving transcript path graph for ENST00000409276.5 as figures/SRSF7_novel_ENST00000409276.5_path.png
Saving transcript path graph for ENST00000425778.5 as figures/SRSF7_novel_ENST00000425778.5_path.png
Saving transcript path graph for ENST00000431066.5 as figures/SRSF7_novel_ENST00000431066.5_path.png
Saving transcript path graph for ENST00000477635.5 as figures/SRSF7_novel_ENST00000477635.5_path.png
Generating report for ENSG00000115875.18

Plotting transcripts for ENSG00000115875.18
Saving transcript path graph for ENST00000313117.10 as figures/SRSF7_dpi_browser_ENST00000313117.10_path.png
Saving transcript path graph for ENST00000446327.6 as figures/SRSF7_dpi_browser_ENST00000446327.6_path.png
Saving transcript path graph for ENST00000443213.5 as figures/SRSF7_dpi_browser_ENST00000443213.5_path.png
Saving transcript path graph for ENST00000409276.5 as figures/SRSF7_dpi_browser_ENST00000409276.5_path.png
Saving transcript path graph for ENST00000425778.5 as figures/SRSF7_dpi_browser_ENST00000425778.5_path.png
Saving transcript path graph for ENST00000431066.5 as figures/SRSF7_dpi_browser_ENST00000431066.5_path.png
Saving transcript path graph for ENST00000477635.5 as figures/SRSF7_dpi_browser_ENST00000477635.5_path.png
Generating report for ENSG00000115875.18

Plotting transcripts for ENSG00000115875.18
Saving transcript path graph for ENST00000313117.10 as figures/SRSF7_browser_ENST00000313117.10_path.png
Saving transcript path graph for ENST00000446327.6 as figures/SRSF7_browser_ENST00000446327.6_path.png
Saving transcript path graph for ENST00000443213.5 as figures/SRSF7_browser_ENST00000443213.5_path.png
Saving transcript path graph for ENST00000409276.5 as figures/SRSF7_browser_ENST00000409276.5_path.png
Saving transcript path graph for ENST00000425778.5 as figures/SRSF7_browser_ENST00000425778.5_path.png
Saving transcript path graph for ENST00000431066.5 as figures/SRSF7_browser_ENST00000431066.5_path.png
Saving transcript path graph for ENST00000477635.5 as figures/SRSF7_browser_ENST00000477635.5_path.png
Generating report for ENSG00000115875.18
</code></pre></div></div>

<p>TPD52L2 reports: (Forgive the terrible sample labels, this will be fixed in a future Swan release!)</p>

<p><img src="/assets/images/long-read/TPD52L2_swan.png" alt="png" /></p>

<p><img src="/assets/images/long-read/TPD52L2_browser.png" alt="png" /></p>

<p>AP2M1 reports: (Forgive the terrible sample labels, this will be fixed in a future Swan release!)</p>

<p><img src="/assets/images/long-read/AP2M1_swan.png" alt="png" /></p>

<p><img src="/assets/images/long-read/AP2M1_browser.png" alt="png" /></p>

<h2 id="visualizing-reads-and-tsss-on-the-genome-browser">Visualizing reads and TSSs on the genome browser</h2>

<h3 id="hub-file">Hub file</h3>
<p>Use <a href="http://crick.bio.uci.edu/freese/210413_pgp1_hub/hub/hub.txt">this link</a> to view my tracks on your genome browser session</p>

<h3 id="convert-tss-bed-to-bigbed">Convert TSS bed to bigBed</h3>
<p>Convert the bed file of filtered TSSs that can be displayed on the genome browser.</p>

<p>To download the bedToBigBed software, use the conda installation instructions here: https://anaconda.org/bioconda/ucsc-bedtobigbed</p>

<div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">bed</span><span class="o">=</span><span class="s1">'filt_tss.bed'</span>
<span class="nv">chrom_sizes</span><span class="o">=</span><span class="s1">'hg38.chrom.sizes'</span>
bedToBigBed <span class="se">\</span>
    <span class="nv">$bed</span> <span class="se">\</span>
    <span class="nv">$chrom_sizes</span> <span class="se">\</span>
    filt_tss.bb
</code></pre></div></div>

<h3 id="creating-and-hosting-the-track-hub">Creating and hosting the track hub</h3>

<p>This part requires you have use of a server with public-facing access <a href="http://crick.bio.uci.edu/freese/">example</a>. The old HPC had one which was really useful (the link used to look like https://hpc.oit.uci.edu/~freese), but there is none for HPC3 and they are not planning on releasing one (I asked HPC support about this). Therefore I’m not sure how useful this will be for those that do not have their own lab server.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">c_dict</span><span class="p">,</span> <span class="n">order</span> <span class="o">=</span> <span class="n">get_sample_colors</span><span class="p">()</span>
<span class="n">sample_file</span> <span class="o">=</span> <span class="s">'samples.txt'</span> <span class="c1"># sample file with one sample per line
</span><span class="n">genome</span> <span class="o">=</span> <span class="s">'hg38'</span> <span class="c1"># genome
</span><span class="n">hub_name</span> <span class="o">=</span> <span class="s">'pgp1'</span> <span class="c1"># name of hub
</span><span class="n">tss</span> <span class="o">=</span> <span class="s">'filt_tss.bb'</span> <span class="c1"># tss file
</span><span class="n">email</span> <span class="o">=</span> <span class="s">'freese@uci.edu'</span> <span class="c1"># email of hub creator
</span>
<span class="c1"># URL location of hub
</span><span class="n">url</span> <span class="o">=</span> <span class="s">'http://crick.bio.uci.edu/freese/210413_pgp1_hub/'</span>

<span class="c1"># ssh location of hub (used to copy)
# if you do not want to automatically move files to server, use the
# commented-out version of the `make_hub()` call.
</span><span class="n">scp_location</span> <span class="o">=</span> <span class="s">'freese@crick.bio.uci.edu:~/pub/210413_pgp1_hub/'</span>

<span class="n">make_hub</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">hub_name</span><span class="p">,</span> <span class="n">email</span><span class="p">,</span> <span class="n">scp_location</span><span class="o">=</span><span class="n">scp_location</span><span class="p">)</span>
<span class="c1"># make_hub(url, c_dict, samples, genome, hub_name, email, scp_location=None)
</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">make_bam_hub_entry</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">ofile</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">ofile</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">df</span><span class="p">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">c_dict</span><span class="p">[</span><span class="n">e</span><span class="p">[</span><span class="s">'sample'</span><span class="p">]]</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">c</span><span class="p">.</span><span class="n">lstrip</span><span class="p">(</span><span class="s">'#'</span><span class="p">)</span>
            <span class="n">c</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">c</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">],</span> <span class="mi">16</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

            <span class="n">s</span> <span class="o">=</span> <span class="s">'track {}_reads</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">e</span><span class="p">[</span><span class="s">'sample'</span><span class="p">])</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'bigDataUrl {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">e</span><span class="p">.</span><span class="n">url</span><span class="p">)</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'shortLabel {}_reads</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">e</span><span class="p">[</span><span class="s">'sample'</span><span class="p">])</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'longLabel {}_reads</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">e</span><span class="p">[</span><span class="s">'sample'</span><span class="p">])</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'type bam</span><span class="se">\n</span><span class="s">'</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'visibility squish</span><span class="se">\n</span><span class="s">'</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'bamColorMode off</span><span class="se">\n</span><span class="s">'</span>
            <span class="n">s</span> <span class="o">+=</span> <span class="s">'color {},{},{}</span><span class="se">\n\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">c</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">c</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="n">c</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="n">o</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">make_tss_hub_entry</span><span class="p">(</span><span class="n">tss</span><span class="p">,</span> <span class="n">url</span><span class="p">,</span> <span class="n">ofile</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">ofile</span><span class="p">,</span> <span class="s">'a'</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s">'track tss</span><span class="se">\n</span><span class="s">'</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'type bigBed 6</span><span class="se">\n</span><span class="s">'</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'bigDataUrl {}{}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">tss</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'shortLabel tss </span><span class="se">\n</span><span class="s">'</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'longLabel tss</span><span class="se">\n</span><span class="s">'</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'visibility dense</span><span class="se">\n</span><span class="s">'</span>
        <span class="n">o</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="c1"># this function probably won't be useful for anyone
# relies on talon_tmp/ bam files
</span><span class="k">def</span> <span class="nf">make_hub</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">samples</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">hub_name</span><span class="p">,</span> <span class="n">email</span><span class="p">,</span> <span class="n">scp_location</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>

    <span class="n">genomefile</span> <span class="o">=</span> <span class="s">'hub/genomes.txt'</span>
    <span class="n">hubfile</span> <span class="o">=</span> <span class="s">'hub/hub.txt'</span>
    <span class="n">hubfile_relative</span> <span class="o">=</span> <span class="s">'hub.txt'</span>
    <span class="n">trackdb</span> <span class="o">=</span> <span class="s">'hub/{}/trackDb.txt'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="n">relative_trackdb</span> <span class="o">=</span> <span class="s">'{}/trackDb.txt'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
    <span class="n">relative_genome</span> <span class="o">=</span> <span class="s">'genomes.txt'</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">os</span><span class="p">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="p">.</span><span class="n">path</span><span class="p">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">trackdb</span><span class="p">))</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="p">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s">'sample'</span><span class="p">])</span>

    <span class="n">df</span><span class="p">[</span><span class="s">'url'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="nb">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">url</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">'sample'</span><span class="p">]</span><span class="o">+</span><span class="s">'.bam'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s">'local_loc'</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">.</span><span class="nb">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s">'talon_tmp/'</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s">'sample'</span><span class="p">]</span><span class="o">+</span><span class="s">'.bam'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">ind</span><span class="p">,</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">df</span><span class="p">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="s">'samtools index {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">entry</span><span class="p">.</span><span class="n">local_loc</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="n">shlex</span><span class="p">.</span><span class="n">split</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="p">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">scp_location</span><span class="p">:</span>
            <span class="n">cmd</span> <span class="o">=</span> <span class="s">'scp {} {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">entry</span><span class="p">.</span><span class="n">local_loc</span><span class="p">,</span> <span class="n">scp_location</span><span class="p">)</span>
            <span class="n">cmd</span> <span class="o">=</span> <span class="n">shlex</span><span class="p">.</span><span class="n">split</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
            <span class="k">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="p">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

            <span class="n">cmd</span> <span class="o">=</span> <span class="s">'scp {}.bai {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">entry</span><span class="p">.</span><span class="n">local_loc</span><span class="p">,</span> <span class="n">scp_location</span><span class="p">)</span>
            <span class="n">cmd</span> <span class="o">=</span> <span class="n">shlex</span><span class="p">.</span><span class="n">split</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
            <span class="k">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="p">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>

    <span class="n">make_bam_hub_entry</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">c_dict</span><span class="p">,</span> <span class="n">trackdb</span><span class="p">)</span>
    <span class="n">make_tss_hub_entry</span><span class="p">(</span><span class="n">tss</span><span class="p">,</span> <span class="n">url</span><span class="p">,</span> <span class="n">trackdb</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genomefile</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s">'genome {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'trackDb {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">relative_trackdb</span><span class="p">)</span>
        <span class="n">o</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">hubfile</span><span class="p">,</span> <span class="s">'w'</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">s</span> <span class="o">=</span> <span class="s">'hub {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">hub_name</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'shortLabel {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">hub_name</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'longLabel {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">hub_name</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'genomesFile {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">relative_genome</span><span class="p">)</span>
        <span class="n">s</span> <span class="o">+=</span> <span class="s">'email {}</span><span class="se">\n</span><span class="s">'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">email</span><span class="p">)</span>
        <span class="n">o</span><span class="p">.</span><span class="n">write</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">scp_location</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="s">'scp -r hub/ {}'</span><span class="p">.</span><span class="nb">format</span><span class="p">(</span><span class="n">scp_location</span><span class="p">)</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="n">shlex</span><span class="p">.</span><span class="n">split</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="p">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</code></pre></div></div>
<p>A (cropped) view of the TPD52L2 locus on the track hub, with TSSs in black, astrocyte reads in orange, PGP1 reads in blue, and excitatory neuron reads in pink. Blue highlights (added afterward) show locations of alternative exons</p>

<p><img src="/assets/images/long-read/track_hub.png" alt="png" /></p>


    </div>

</article>
<div class="post-nav"><a class="previous" href="/hpc/2021/04/07/CondaEnvOnHPC3.html" title="Python on HPC3">Python on HPC3</a><a class="next" href="/splicing/2022/01/27/swan.html" title="Swan">Swan</a></div><div class="post-related">
      <div>Related Articles</div>
      <ul>
        <li><a class="post-link" href="/network/2022/02/24/scWGCNA.html" title="Swan">scWGCNA</a></li><li><a class="post-link" href="/splicing/2022/01/27/swan.html" title="Swan">Swan</a></li><li><a class="post-link" href="/long-read/2021/04/19/long-reads.html" title="Swan">Analysis of long-read transcriptomes</a></li><li><a class="post-link" href="/preprocessing/2021/04/05/Tale-of-Two-Aligners.html" title="Swan">A Tale of Two Aligners</a></li></ul>
    </div><div class="post-comments">  <div id="disqus_thread"></div>
  <script>
    var disqus_config = function () {
      this.page.url = '/long-read/2021/04/19/long-reads.html';
      this.page.identifier = '/long-read/2021/04/19/long-reads.html';
    };

    (function() {
      var d = document, s = d.createElement('script');

      s.src = 'https://https-uci-genpals-github-io.disqus.com/embed.js';

      s.setAttribute('data-timestamp', +new Date());
      (d.head || d.body).appendChild(s);
    })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>
</div></section>
</div>


  </section>
  <section class="sidebar" style="margin-left: 15px;">
    <!-- Get sidebar items --></section>
</div>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">
    <div class="site-footer-inner">
      <div>Copyright © 2020-2021 @UCI genPALS</div>
      <div>Powered by <a title="Jekyll is a simple, blog-aware, static site
      generator." href="http://jekyllrb.com/">Jekyll</a> &amp; <a title="Yat, yet
      another theme." href="https://github.com/jeffreytse/jekyll-theme-yat">Yat Theme</a>.</div>
      <div class="footer-col rss-subscribe">Subscribe <a href="/feed.xml">via RSS</a></div>
    </div>
  </div>
</footer>
</body>
</html>
